{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c1908fc4-328c-4ba8-a472-27cd6b1a0df1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import oakvar as ov"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "045df754-bc28-4d85-9008-b277aba8d72d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Install and setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947b7a3a-df66-441a-aa9f-0f3aa7a10470",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4841df34-520c-4875-a873-de638ff6be17",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -U oakvar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a935c03-7bb8-4b35-a642-ea7834489561",
   "metadata": {},
   "source": [
    "##### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d6627bd-df32-4257-b550-935ce2b27e46",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ov.system.setup(outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37e41fed-8ccf-4ab9-8d37-4de81ae0ba45",
   "metadata": {},
   "source": [
    "### Login and logout"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24293dd-0870-471c-886f-7678149672c2",
   "metadata": {},
   "source": [
    "_Check login state_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e42e60e2-883d-44af-9998-6a1341f833c0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "ov.store.account.check()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec7b974a-b680-44a0-9e55-45d030cc9f12",
   "metadata": {},
   "source": [
    "_Logout_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b80361-1818-4fe2-b82b-26949ae2f3b2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "ov.store.logout(outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6d9de6e-0669-46d2-8412-e79b3fd67063",
   "metadata": {},
   "source": [
    "_Login with email and password arguments_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37698192-7e5f-4ecc-b2e2-4160172186a2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "ov.store.login(email=\"your@email\", pw=\"your_password\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1c1e395-3c8c-4ca8-af57-379aa0156122",
   "metadata": {},
   "source": [
    "_Login interactively_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afbdb800-f9e8-4e53-a244-09bc4486f347",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "ov.store.login(interactive=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ce35e7-7fd8-4d89-94f1-2cedc8e87203",
   "metadata": {},
   "source": [
    "_Change password_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ac4b96-0119-441d-9e81-072a658e9db8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "ov.store.account.change(outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d46a2ed6-a8f1-4918-a5a7-6af469ac878b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Module management"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947575c9-5208-4741-8365-3de73a303fbf",
   "metadata": {},
   "source": [
    "_List installed modules_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "25c7695f-7112-44de-bc17-43f198aaa077",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': '23andme-converter',\n",
       "  'title': '23andMe Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '21.9 kB',\n",
       "  'version': '1.5.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'abraom',\n",
       "  'title': 'ABRaOM',\n",
       "  'type': 'annotator',\n",
       "  'size': '113.6 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'af-package',\n",
       "  'title': 'Allele Frequency Package',\n",
       "  'type': 'package',\n",
       "  'size': '881 B',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'af2-package',\n",
       "  'title': 'Allele Frequency Package',\n",
       "  'type': 'package',\n",
       "  'size': '790 B',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'aggregator',\n",
       "  'title': 'Aggregator',\n",
       "  'type': 'aggregator',\n",
       "  'size': '32.5 kB',\n",
       "  'version': '1.0.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'alfa',\n",
       "  'title': 'ALFA: Allele Frequency Aggregator',\n",
       "  'type': 'annotator',\n",
       "  'size': '2.4 GB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '2020.02.29'},\n",
       " {'name': 'alfa_african',\n",
       "  'title': 'ALFA: Allele Frequency Aggregator African',\n",
       "  'type': 'annotator',\n",
       "  'size': '1.9 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '2020.02.29'},\n",
       " {'name': 'alfa_asian',\n",
       "  'title': 'ALFA: Allele Frequency Aggregator Asian',\n",
       "  'type': 'annotator',\n",
       "  'size': '5.4 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '2020.02.29'},\n",
       " {'name': 'alfa_european',\n",
       "  'title': 'ALFA: Allele Frequency Aggregator European',\n",
       "  'type': 'annotator',\n",
       "  'size': '53.1 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '2020.02.29'},\n",
       " {'name': 'alfa_group',\n",
       "  'title': 'ALFA: Allele Frequency Aggregator',\n",
       "  'type': 'group',\n",
       "  'size': '101.7 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '2020.02.29'},\n",
       " {'name': 'aloft',\n",
       "  'title': 'ALoFT',\n",
       "  'type': 'annotator',\n",
       "  'size': '260.4 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v4.1a'},\n",
       " {'name': 'ancestrydna-converter',\n",
       "  'title': 'AncestryDNA Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '33.8 kB',\n",
       "  'version': '1.4.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'arrvars',\n",
       "  'title': 'Arrhythmia Channelopathy Variants',\n",
       "  'type': 'annotator',\n",
       "  'size': '2.1 MB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': 'January 2020'},\n",
       " {'name': 'biogrid',\n",
       "  'title': 'BioGRID',\n",
       "  'type': 'annotator',\n",
       "  'size': '78.1 MB',\n",
       "  'version': '3.7.1',\n",
       "  'data_source': 'v4.4.197'},\n",
       " {'name': 'brca1_func_assay',\n",
       "  'title': 'BRCA1 Saturation Genome Editing Scores',\n",
       "  'type': 'annotator',\n",
       "  'size': '263.5 kB',\n",
       "  'version': '2019.05.13',\n",
       "  'data_source': '2018.04.05'},\n",
       " {'name': 'cadd',\n",
       "  'title': 'CADD',\n",
       "  'type': 'annotator',\n",
       "  'size': '86.6 GB',\n",
       "  'version': '1.6.0',\n",
       "  'data_source': 'v1.6'},\n",
       " {'name': 'cancer_genome_interpreter',\n",
       "  'title': 'Cancer Genome Interpreter',\n",
       "  'type': 'annotator',\n",
       "  'size': '554.9 kB',\n",
       "  'version': '1.3.0',\n",
       "  'data_source': '2018.01.17'},\n",
       " {'name': 'cancer_hotspots',\n",
       "  'title': 'Cancer Hotspots',\n",
       "  'type': 'annotator',\n",
       "  'size': '785.9 kB',\n",
       "  'version': '1.2.1',\n",
       "  'data_source': 'December 2017'},\n",
       " {'name': 'cardioboost',\n",
       "  'title': 'CardioBoost',\n",
       "  'type': 'annotator',\n",
       "  'size': '18.2 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v2.1'},\n",
       " {'name': 'casecontrol',\n",
       "  'title': 'Case-Control',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '38.4 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'ccr',\n",
       "  'title': 'CCR: Constrained Coding Regions',\n",
       "  'type': 'annotator',\n",
       "  'size': '8.0 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'December 2018'},\n",
       " {'name': 'ccre_screen',\n",
       "  'title': 'Candidate cis-Regulatory Elements by ENCODE (SCREEN)',\n",
       "  'type': 'annotator',\n",
       "  'size': '77.8 MB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': 2019.1},\n",
       " {'name': 'cgc',\n",
       "  'title': 'Cancer Gene Census',\n",
       "  'type': 'annotator',\n",
       "  'size': '766.5 kB',\n",
       "  'version': '85.0.12',\n",
       "  'data_source': '2019-03-23'},\n",
       " {'name': 'cgd',\n",
       "  'title': 'CGD: Clinical Genomic Database',\n",
       "  'type': 'annotator',\n",
       "  'size': '706.0 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': '2021.02.09'},\n",
       " {'name': 'cgl',\n",
       "  'title': 'Cancer Gene Landscape',\n",
       "  'type': 'annotator',\n",
       "  'size': '998.7 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': 'v2013.03.29'},\n",
       " {'name': 'civic',\n",
       "  'title': 'CIViC',\n",
       "  'type': 'annotator',\n",
       "  'size': '592.0 kB',\n",
       "  'version': '1.0.16',\n",
       "  'data_source': 'Jan 2019'},\n",
       " {'name': 'civic_gene',\n",
       "  'title': 'CIViC Gene',\n",
       "  'type': 'annotator',\n",
       "  'size': '55.7 kB',\n",
       "  'version': '1.0.10',\n",
       "  'data_source': 'Jan 2019'},\n",
       " {'name': 'clingen',\n",
       "  'title': 'ClinGen Gene',\n",
       "  'type': 'annotator',\n",
       "  'size': '190.4 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'clinvar',\n",
       "  'title': 'ClinVar',\n",
       "  'type': 'annotator',\n",
       "  'size': '340.6 MB',\n",
       "  'version': '2022.06.14.1',\n",
       "  'data_source': '2022.06.14'},\n",
       " {'name': 'clinvar_acmg',\n",
       "  'title': 'ClinVar ACMG',\n",
       "  'type': 'annotator',\n",
       "  'size': '67.1 MB',\n",
       "  'version': '2021.10.01',\n",
       "  'data_source': '2021.10.01'},\n",
       " {'name': 'cosmic',\n",
       "  'title': 'COSMIC',\n",
       "  'type': 'annotator',\n",
       "  'size': '1.6 GB',\n",
       "  'version': '94.0.0',\n",
       "  'data_source': 'v94'},\n",
       " {'name': 'cosmic_gene',\n",
       "  'title': 'COSMIC Gene',\n",
       "  'type': 'annotator',\n",
       "  'size': '134.1 MB',\n",
       "  'version': '94.0.0',\n",
       "  'data_source': 'v94'},\n",
       " {'name': 'cravat-converter',\n",
       "  'title': 'Cravat Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '15.0 kB',\n",
       "  'version': '1.1.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'cscape',\n",
       "  'title': 'CScape',\n",
       "  'type': 'annotator',\n",
       "  'size': '588.2 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': 2017},\n",
       " {'name': 'cscape_coding',\n",
       "  'title': 'CScape Coding',\n",
       "  'type': 'annotator',\n",
       "  'size': '5.2 GB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 2017},\n",
       " {'name': 'csvreporter',\n",
       "  'title': 'CSV Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '270.4 kB',\n",
       "  'version': '2.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'dann',\n",
       "  'title': 'DANN',\n",
       "  'type': 'annotator',\n",
       "  'size': '109.6 GB',\n",
       "  'version': '1.0.2.1',\n",
       "  'data_source': 'March 2015'},\n",
       " {'name': 'dbcid',\n",
       "  'title': 'dbCID: Database of Cancer Driver InDels',\n",
       "  'type': 'annotator',\n",
       "  'size': '206.3 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': 'v1.0'},\n",
       " {'name': 'dbscsnv',\n",
       "  'title': 'dbscSNV',\n",
       "  'type': 'annotator',\n",
       "  'size': '754.9 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v1.1'},\n",
       " {'name': 'dbsnp',\n",
       "  'title': 'dbSNP',\n",
       "  'type': 'annotator',\n",
       "  'size': '47.1 GB',\n",
       "  'version': '155.0.1',\n",
       "  'data_source': 155},\n",
       " {'name': 'denovo',\n",
       "  'title': 'Denovo-DB',\n",
       "  'type': 'annotator',\n",
       "  'size': '39.5 MB',\n",
       "  'version': '1.6.11',\n",
       "  'data_source': 'v161'},\n",
       " {'name': 'dfpostagg',\n",
       "  'title': 'Polars postaggregator example',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '2.0 kB',\n",
       "  'version': '0.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'dgi',\n",
       "  'title': 'DGIdb: The Drug Interaction Database',\n",
       "  'type': 'annotator',\n",
       "  'size': '7.5 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'Jan 2021'},\n",
       " {'name': 'dida',\n",
       "  'title': 'DIDA: Digenic Diseases Database',\n",
       "  'type': 'annotator',\n",
       "  'size': '242.2 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'July 2017'},\n",
       " {'name': 'encode_tfbs',\n",
       "  'title': 'ENCODE TFBS',\n",
       "  'type': 'annotator',\n",
       "  'size': '578.6 MB',\n",
       "  'version': '1.0.2',\n",
       "  'data_source': '2012-03'},\n",
       " {'name': 'ensembl_regulatory_build',\n",
       "  'title': 'Ensembl Regulatory Build',\n",
       "  'type': 'annotator',\n",
       "  'size': '49.6 MB',\n",
       "  'version': '1.0.3',\n",
       "  'data_source': '2020.06.09'},\n",
       " {'name': 'esp6500',\n",
       "  'title': 'ESP6500',\n",
       "  'type': 'annotator',\n",
       "  'size': '116.1 MB',\n",
       "  'version': '6500.2.6',\n",
       "  'data_source': 'v0.0.30'},\n",
       " {'name': 'ess_gene',\n",
       "  'title': 'Essential Genes',\n",
       "  'type': 'annotator',\n",
       "  'size': '874.6 kB',\n",
       "  'version': '3.5.8',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'exac_gene',\n",
       "  'title': 'ExAC Gene and CNV',\n",
       "  'type': 'annotator',\n",
       "  'size': '2.5 MB',\n",
       "  'version': '0.4.0',\n",
       "  'data_source': 'v0.3.1'},\n",
       " {'name': 'excelreporter',\n",
       "  'title': 'Excel Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '111.8 kB',\n",
       "  'version': '2.2.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'fathmm',\n",
       "  'title': 'FATHMM',\n",
       "  'type': 'annotator',\n",
       "  'size': '10.4 GB',\n",
       "  'version': '2.3.7.1',\n",
       "  'data_source': 'v2.3'},\n",
       " {'name': 'fathmm_xf_coding',\n",
       "  'title': 'FATHMM XF Coding',\n",
       "  'type': 'annotator',\n",
       "  'size': '4.1 GB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v4.1'},\n",
       " {'name': 'ftdna-converter',\n",
       "  'title': 'FamilyTreeDNA Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '35.3 kB',\n",
       "  'version': '1.4.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'gencode',\n",
       "  'title': 'GENCODE Mapper',\n",
       "  'type': 'mapper',\n",
       "  'size': '540.6 MB',\n",
       "  'version': '41.3.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'genehancer',\n",
       "  'title': 'GeneHancer',\n",
       "  'type': 'annotator',\n",
       "  'size': '42.4 MB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': 'v4.7'},\n",
       " {'name': 'geuvadis',\n",
       "  'title': 'Geuvadis eQTLs',\n",
       "  'type': 'annotator',\n",
       "  'size': '289.8 kB',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'ghis',\n",
       "  'title': 'GHIS',\n",
       "  'type': 'annotator',\n",
       "  'size': '701.5 kB',\n",
       "  'version': '3.6.0',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'gnomad',\n",
       "  'title': 'gnomAD',\n",
       "  'type': 'annotator',\n",
       "  'size': '17.4 GB',\n",
       "  'version': '2.2.0',\n",
       "  'data_source': 'v2.1'},\n",
       " {'name': 'gnomad_gene',\n",
       "  'title': 'gnomAD Gene',\n",
       "  'type': 'annotator',\n",
       "  'size': '11.1 MB',\n",
       "  'version': '2.2.1',\n",
       "  'data_source': 'v2.1'},\n",
       " {'name': 'go',\n",
       "  'title': 'Gene Ontology',\n",
       "  'type': 'annotator',\n",
       "  'size': '33.2 MB',\n",
       "  'version': '2022.10.05',\n",
       "  'data_source': '2022.10.05'},\n",
       " {'name': 'grasp',\n",
       "  'title': 'GRASP',\n",
       "  'type': 'annotator',\n",
       "  'size': '715.0 MB',\n",
       "  'version': '2.0.8',\n",
       "  'data_source': 'v2.0.0.0'},\n",
       " {'name': 'gtex',\n",
       "  'title': 'GTEx eQTLs',\n",
       "  'type': 'annotator',\n",
       "  'size': '14.7 MB',\n",
       "  'version': '7.0.4',\n",
       "  'data_source': 'v7'},\n",
       " {'name': 'gwas_catalog',\n",
       "  'title': 'GWAS Catalog',\n",
       "  'type': 'annotator',\n",
       "  'size': '39.0 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'haploreg_group',\n",
       "  'title': 'HaploReg',\n",
       "  'type': 'group',\n",
       "  'size': '4.3 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v4.1'},\n",
       " {'name': 'hg19',\n",
       "  'title': 'hg19 coordinates',\n",
       "  'type': 'annotator',\n",
       "  'size': '3.9 MB',\n",
       "  'version': '1.0.6',\n",
       "  'data_source': ''},\n",
       " {'name': 'hg19wgs',\n",
       "  'title': 'UCSC hg19 Whole Genome Sequence',\n",
       "  'type': 'common',\n",
       "  'size': '816.2 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'hg38wgs',\n",
       "  'title': 'UCSC hg38 Whole Genome Sequence',\n",
       "  'type': 'common',\n",
       "  'size': '851.8 MB',\n",
       "  'version': '2020.03.13',\n",
       "  'data_source': '2020.03.13'},\n",
       " {'name': 'hgdp',\n",
       "  'title': 'HGDP',\n",
       "  'type': 'annotator',\n",
       "  'size': '63.4 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'hpo',\n",
       "  'title': 'Human Phenotype Ontology',\n",
       "  'type': 'annotator',\n",
       "  'size': '14.6 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'April 2021'},\n",
       " {'name': 'intact',\n",
       "  'title': 'IntAct',\n",
       "  'type': 'annotator',\n",
       "  'size': '8.8 MB',\n",
       "  'version': '2019.08.23',\n",
       "  'data_source': '2018.11.30'},\n",
       " {'name': 'javierre_promoters',\n",
       "  'title': 'Promoter IR',\n",
       "  'type': 'annotator',\n",
       "  'size': '553.2 kB',\n",
       "  'version': '2019.05.20',\n",
       "  'data_source': 'November 17, 2016'},\n",
       " {'name': 'jsonpost',\n",
       "  'title': 'JSON post',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '1.6 kB',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'just_coronary',\n",
       "  'title': 'Coronary disease postagregator',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '209.6 kB',\n",
       "  'version': '0.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'litvar',\n",
       "  'title': 'LitVar',\n",
       "  'type': 'annotator',\n",
       "  'size': '364.9 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': 'Live'},\n",
       " {'name': 'loftool',\n",
       "  'title': 'LoFtool',\n",
       "  'type': 'annotator',\n",
       "  'size': '1.1 MB',\n",
       "  'version': '3.6.1',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'longevity2reporter',\n",
       "  'title': 'Longevity2',\n",
       "  'type': 'reporter',\n",
       "  'size': '271.0 kB',\n",
       "  'version': '0.3.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'longevity_combinedreporter',\n",
       "  'title': 'Longevity Combined',\n",
       "  'type': 'reporter',\n",
       "  'size': '3.1 MB',\n",
       "  'version': '0.1.4',\n",
       "  'data_source': ''},\n",
       " {'name': 'longevitymap',\n",
       "  'title': 'Longevitymap postagregator',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '3.0 MB',\n",
       "  'version': '0.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'mavedb',\n",
       "  'title': 'MaveDB',\n",
       "  'type': 'annotator',\n",
       "  'size': '2.1 MB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': 'v1.7.2'},\n",
       " {'name': 'mitomap',\n",
       "  'title': 'MITOMAP',\n",
       "  'type': 'annotator',\n",
       "  'size': '376.9 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': '2020.04.27'},\n",
       " {'name': 'moleculartumorboard',\n",
       "  'title': 'Molecular Tumor Board Variant Report',\n",
       "  'type': 'webapp',\n",
       "  'size': '7.2 MB',\n",
       "  'version': '1.2.4',\n",
       "  'data_source': ''},\n",
       " {'name': 'mutationtaster',\n",
       "  'title': 'MutationTaster',\n",
       "  'type': 'annotator',\n",
       "  'size': '12.1 GB',\n",
       "  'version': '1.0.0.1',\n",
       "  'data_source': 'v4.1a'},\n",
       " {'name': 'mutpanning',\n",
       "  'title': 'Mutpanning',\n",
       "  'type': 'annotator',\n",
       "  'size': '397.3 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v2.0'},\n",
       " {'name': 'mutpred_indel',\n",
       "  'title': 'MutPred-Indel',\n",
       "  'type': 'annotator',\n",
       "  'size': '17.9 GB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': 'v1.0'},\n",
       " {'name': 'ncbigene',\n",
       "  'title': 'NCBI Gene',\n",
       "  'type': 'annotator',\n",
       "  'size': '20.9 MB',\n",
       "  'version': '2022.08.11',\n",
       "  'data_source': '2022.08.11'},\n",
       " {'name': 'ncrna',\n",
       "  'title': 'ncRNA',\n",
       "  'type': 'annotator',\n",
       "  'size': '5.8 MB',\n",
       "  'version': '2019.08.19',\n",
       "  'data_source': 'hg38'},\n",
       " {'name': 'ndex',\n",
       "  'title': 'NDEx NCI Cancer Pathways',\n",
       "  'type': 'annotator',\n",
       "  'size': '797.4 kB',\n",
       "  'version': '4.1.0',\n",
       "  'data_source': 'v2'},\n",
       " {'name': 'ndex_group',\n",
       "  'title': 'NDEx',\n",
       "  'type': 'group',\n",
       "  'size': '74.8 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 2020.08},\n",
       " {'name': 'oldcravat-converter',\n",
       "  'title': 'OldCravat Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '14.4 kB',\n",
       "  'version': '1.1.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'omim',\n",
       "  'title': 'OMIM',\n",
       "  'type': 'annotator',\n",
       "  'size': '10.4 MB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v4.1'},\n",
       " {'name': 'oncokb',\n",
       "  'title': 'OncoKB',\n",
       "  'type': 'annotator',\n",
       "  'size': '243.4 kB',\n",
       "  'version': '1.2.3',\n",
       "  'data_source': '09.06.22'},\n",
       " {'name': 'pandasreporter',\n",
       "  'title': 'pandas Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '197.9 kB',\n",
       "  'version': '1.2.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'pharmgkb',\n",
       "  'title': 'PharmGKB',\n",
       "  'type': 'annotator',\n",
       "  'size': '3.7 MB',\n",
       "  'version': '2.3.5',\n",
       "  'data_source': '2021-10-01'},\n",
       " {'name': 'phastcons',\n",
       "  'title': 'Phast Cons',\n",
       "  'type': 'annotator',\n",
       "  'size': '6.3 GB',\n",
       "  'version': '3.5.9',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'phi',\n",
       "  'title': 'p(HI)',\n",
       "  'type': 'annotator',\n",
       "  'size': '711.0 kB',\n",
       "  'version': '3.6.1',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'phylop',\n",
       "  'title': 'PhyloP',\n",
       "  'type': 'annotator',\n",
       "  'size': '284.3 MB',\n",
       "  'version': '3.5.10',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'polyphen2',\n",
       "  'title': 'PolyPhen-2',\n",
       "  'type': 'annotator',\n",
       "  'size': '31.3 kB',\n",
       "  'version': '4.3.0.1',\n",
       "  'data_source': 'v4.3.0.1'},\n",
       " {'name': 'prs',\n",
       "  'title': 'PRS Annotator',\n",
       "  'type': 'annotator',\n",
       "  'size': '6.1 kB',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'pubmed',\n",
       "  'title': 'PubMed',\n",
       "  'type': 'annotator',\n",
       "  'size': '989.5 kB',\n",
       "  'version': '1.1.5',\n",
       "  'data_source': ''},\n",
       " {'name': 'rdatareporter',\n",
       "  'title': 'RData Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '17.9 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'repeat',\n",
       "  'title': 'Repeat Sequences',\n",
       "  'type': 'annotator',\n",
       "  'size': '366.4 MB',\n",
       "  'version': '2020.10.16',\n",
       "  'data_source': 'hg38'},\n",
       " {'name': 'segway_gi_duodenum',\n",
       "  'title': 'Segway GI Duodenum',\n",
       "  'type': 'annotator',\n",
       "  'size': '3.7 MB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': '2019.09.01'},\n",
       " {'name': 'siphy',\n",
       "  'title': 'SiPhy',\n",
       "  'type': 'annotator',\n",
       "  'size': '5.6 GB',\n",
       "  'version': '3.5.5',\n",
       "  'data_source': 'dbNSFP v4.0b1'},\n",
       " {'name': 'spliceai',\n",
       "  'title': 'SpliceAI',\n",
       "  'type': 'annotator',\n",
       "  'size': '1.8 GB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': 'v4.0'},\n",
       " {'name': 'stdoutreporter',\n",
       "  'title': 'Standard Output Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '2.7 kB',\n",
       "  'version': '1.2.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'svracas',\n",
       "  'title': 'SVRACAS',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '26.0 kB',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'tagsampler',\n",
       "  'title': 'Samples and Tags',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '8.5 kB',\n",
       "  'version': '2.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'testpostaggregator',\n",
       "  'title': 'Samples and Tags',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '9.0 kB',\n",
       "  'version': '2.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'testpreparer',\n",
       "  'title': 'Test Preparer',\n",
       "  'type': 'unknown',\n",
       "  'size': '1.4 kB',\n",
       "  'version': '0.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'testreporter',\n",
       "  'title': 'Test Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '260.3 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'textreporter',\n",
       "  'title': 'Text Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '279.7 kB',\n",
       "  'version': '2.2.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'thousandgenomes',\n",
       "  'title': '1000 Genomes',\n",
       "  'type': 'annotator',\n",
       "  'size': '5.3 GB',\n",
       "  'version': '4.1.0',\n",
       "  'data_source': 'Phase 3'},\n",
       " {'name': 'thousandgenomes_group',\n",
       "  'title': '1000 Genomes',\n",
       "  'type': 'group',\n",
       "  'size': '17.6 kB',\n",
       "  'version': '3.1.0',\n",
       "  'data_source': 'Phase 3'},\n",
       " {'name': 'trinity',\n",
       "  'title': 'Trinity CTAT',\n",
       "  'type': 'annotator',\n",
       "  'size': '157.1 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': '26 February 2020'},\n",
       " {'name': 'tsvreporter',\n",
       "  'title': 'TSV Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '264.1 kB',\n",
       "  'version': '2.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'uniprot',\n",
       "  'title': 'UniProt',\n",
       "  'type': 'annotator',\n",
       "  'size': '774.6 kB',\n",
       "  'version': '2020.08.10',\n",
       "  'data_source': '201809'},\n",
       " {'name': 'variantreport',\n",
       "  'title': 'OpenCRAVAT Variant Report',\n",
       "  'type': 'webapp',\n",
       "  'size': '7.1 MB',\n",
       "  'version': '1.0.7',\n",
       "  'data_source': ''},\n",
       " {'name': 'varmeta',\n",
       "  'title': 'Variant Metadata',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '3.0 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'vcf-converter',\n",
       "  'title': 'VCF Converter',\n",
       "  'type': 'converter',\n",
       "  'size': '50.0 kB',\n",
       "  'version': '2.4.4',\n",
       "  'data_source': ''},\n",
       " {'name': 'vcfinfo',\n",
       "  'title': 'VCF Info',\n",
       "  'type': 'postaggregator',\n",
       "  'size': '12.1 kB',\n",
       "  'version': '1.2.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'vcfreporter',\n",
       "  'title': 'VCF Reporter',\n",
       "  'type': 'reporter',\n",
       "  'size': '294.1 kB',\n",
       "  'version': '4.3.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'vep',\n",
       "  'title': 'VEP annotation module',\n",
       "  'type': 'annotator',\n",
       "  'size': '760 B',\n",
       "  'version': '0.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'vest',\n",
       "  'title': 'VEST4',\n",
       "  'type': 'annotator',\n",
       "  'size': '9.8 GB',\n",
       "  'version': '4.4.0',\n",
       "  'data_source': 'v4'},\n",
       " {'name': 'vista_enhancer',\n",
       "  'title': 'VISTA Enhancer Browser',\n",
       "  'type': 'annotator',\n",
       "  'size': '116.7 kB',\n",
       "  'version': '2019.08.02',\n",
       "  'data_source': 'hg19'},\n",
       " {'name': 'wgallelefrequency',\n",
       "  'title': 'Allele Frequency Spectrum',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '103.2 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgaloft',\n",
       "  'title': 'ALoFT',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.8 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgbase',\n",
       "  'title': 'Variant Annotation',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '4.7 kB',\n",
       "  'version': '1.3.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgbiogrid',\n",
       "  'title': 'BioGRID',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '58.3 kB',\n",
       "  'version': '1.3.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgbrca1_func_assay',\n",
       "  'title': 'BRCA1 Functional Assay',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '729 B',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcancer_genome_interpreter',\n",
       "  'title': 'Cancer Genome Interpreter',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.4 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcancer_hotspots',\n",
       "  'title': 'Cancer Hotspots',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '3.1 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcasecontrols',\n",
       "  'title': 'Case-Control - Most significant p-values',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '74.5 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcasecontrolsummary',\n",
       "  'title': 'Case-Control Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '82.5 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgccre_screen',\n",
       "  'title': 'Candidate cis_Regulatory Elements by ENCODE (SCREEN)',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.6 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcgc',\n",
       "  'title': 'Cancer Genome Census',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '15.0 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcgd',\n",
       "  'title': 'CGD: Clinical Genomic Database',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.5 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcircossummary',\n",
       "  'title': 'Circos Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '898.2 kB',\n",
       "  'version': '2.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcivic',\n",
       "  'title': 'CIViC',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '45.2 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcivic_gene',\n",
       "  'title': 'CIViC Gene',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '45.3 kB',\n",
       "  'version': '1.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgclinvar',\n",
       "  'title': 'Clinvar',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '41.2 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcodingvsnoncodingsummary',\n",
       "  'title': 'Coding vs Noncoding Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '4.7 kB',\n",
       "  'version': '1.4.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgcosmic_gene',\n",
       "  'title': 'Cosmic Gene',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '16.6 kB',\n",
       "  'version': '1.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgdgi',\n",
       "  'title': 'DGIdb',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.6 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgdida',\n",
       "  'title': 'DIDA',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '4.2 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgencode_tfbs',\n",
       "  'title': 'ENCODE TFBS',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.6 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgenhancer',\n",
       "  'title': 'Frequently Mutated Enhancers',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '42.6 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgensembl_regulatory_build',\n",
       "  'title': 'Ensembl Regulatory Build',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.3 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgesp6500',\n",
       "  'title': 'ESP6500',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '17.0 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggenehancer',\n",
       "  'title': 'GeneHancer',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.5 kB',\n",
       "  'version': '1.0.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgghis',\n",
       "  'title': 'GHIS',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '542 B',\n",
       "  'version': '1.0.3',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggnomad',\n",
       "  'title': 'gnomAD',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '23.9 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggnomad_gene',\n",
       "  'title': 'gnomAD Gene',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '24.0 kB',\n",
       "  'version': '1.2.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggo',\n",
       "  'title': 'Gene ontology',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '11.7 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggosummary',\n",
       "  'title': 'Gene Ontology Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '38.8 MB',\n",
       "  'version': '2.6.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggrasp',\n",
       "  'title': 'GRASP',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '7.9 kB',\n",
       "  'version': '1.1.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggtex',\n",
       "  'title': 'GTEx',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '13.0 kB',\n",
       "  'version': '1.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wggwas_catalog',\n",
       "  'title': 'GWAS Catalog',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.6 kB',\n",
       "  'version': '1.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wghg19',\n",
       "  'title': 'hg19 info widget',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '615 B',\n",
       "  'version': '1.0.3',\n",
       "  'data_source': ''},\n",
       " {'name': 'wghpo',\n",
       "  'title': 'Human Phenotype Ontology',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.3 kB',\n",
       "  'version': '1.0.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgintact',\n",
       "  'title': 'IntAct',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '24.1 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wglollipop',\n",
       "  'title': 'Protein diagram',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '187.6 MB',\n",
       "  'version': '2.4.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgncrna',\n",
       "  'title': 'Non-coding RNA diagram',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '50.1 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgndex',\n",
       "  'title': 'NDEx',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '870.1 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgnote',\n",
       "  'title': 'Note',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '4.6 kB',\n",
       "  'version': '2.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgoncokb',\n",
       "  'title': 'OncoKB',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '13.0 kB',\n",
       "  'version': '1.0.2',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgphastcons',\n",
       "  'title': 'Phast Cons',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.1 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgrankscore',\n",
       "  'title': 'Rank Score Comparison',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '64.6 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgsiphy',\n",
       "  'title': 'SiPhy',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '1.5 kB',\n",
       "  'version': '1.1.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgsosamplesummary',\n",
       "  'title': 'Sequence Ontology Sample Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '16.6 kB',\n",
       "  'version': '2.3.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgsosummary',\n",
       "  'title': 'Sequence Ontology Summary',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '6.8 kB',\n",
       "  'version': '1.8.1',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgthousandgenomes',\n",
       "  'title': '1000 Genomes',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '18.0 kB',\n",
       "  'version': '1.1.0',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgvcfinfo',\n",
       "  'title': 'VCF info widget',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '911 B',\n",
       "  'version': '1.0.3',\n",
       "  'data_source': ''},\n",
       " {'name': 'wgvest',\n",
       "  'title': 'VEST',\n",
       "  'type': 'webviewerwidget',\n",
       "  'size': '164.4 kB',\n",
       "  'version': '1.2.0',\n",
       "  'data_source': ''}]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.module.ls()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b776d3d-2794-41ad-9a6e-5eca02d9490a",
   "metadata": {},
   "source": [
    "_List modules available through the OakVar Store. Pattern matching can be used. `outer` is used in this example._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1e9390e6-235c-4026-aaef-f5358ac6132a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">┌────────────┬────────────┬───────────┬──────────┬────────────┬────────────┬───────────┬────────────┬─────────────┐\n",
       "│<span style=\"font-weight: bold\"> name       </span>│<span style=\"font-weight: bold\"> title      </span>│<span style=\"font-weight: bold\"> type      </span>│<span style=\"font-weight: bold\"> size     </span>│<span style=\"font-weight: bold\"> version    </span>│<span style=\"font-weight: bold\"> data_sour… </span>│<span style=\"font-weight: bold\"> installed </span>│<span style=\"font-weight: bold\"> local_cod… </span>│<span style=\"font-weight: bold\"> local_data… </span>│\n",
       "├────────────┼────────────┼───────────┼──────────┼────────────┼────────────┼───────────┼────────────┼─────────────┤\n",
       "│ clingen    │ ClinGen    │ annotator │ 58.0 kB  │ 1.0.1      │ None       │ yes       │ 1.0.1      │             │\n",
       "│            │ Gene       │           │          │            │            │           │            │             │\n",
       "│ clinpred   │ ClinPred   │ annotator │ 941.5 MB │ 1.0.0      │ October    │           │            │             │\n",
       "│            │            │           │          │            │ 2018       │           │            │             │\n",
       "│ clinvar    │ ClinVar    │ annotator │ 45.2 MB  │ 2022.06.14 │ 2022.06.14 │ yes       │ 2022.06.1… │ 2022.06.14  │\n",
       "│ clinvar_a… │ ClinVar    │ annotator │ 14.3 MB  │ 2021.10.01 │ 2021.10.01 │ yes       │ 2021.10.01 │ 2021.10.01  │\n",
       "│            │ ACMG       │           │          │            │            │           │            │             │\n",
       "└────────────┴────────────┴───────────┴──────────┴────────────┴────────────┴───────────┴────────────┴─────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "┌────────────┬────────────┬───────────┬──────────┬────────────┬────────────┬───────────┬────────────┬─────────────┐\n",
       "│\u001b[1m \u001b[0m\u001b[1mname      \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mtitle     \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mtype     \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1msize    \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mversion   \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mdata_sour…\u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1minstalled\u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mlocal_cod…\u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mlocal_data…\u001b[0m\u001b[1m \u001b[0m│\n",
       "├────────────┼────────────┼───────────┼──────────┼────────────┼────────────┼───────────┼────────────┼─────────────┤\n",
       "│ clingen    │ ClinGen    │ annotator │ 58.0 kB  │ 1.0.1      │ None       │ yes       │ 1.0.1      │             │\n",
       "│            │ Gene       │           │          │            │            │           │            │             │\n",
       "│ clinpred   │ ClinPred   │ annotator │ 941.5 MB │ 1.0.0      │ October    │           │            │             │\n",
       "│            │            │           │          │            │ 2018       │           │            │             │\n",
       "│ clinvar    │ ClinVar    │ annotator │ 45.2 MB  │ 2022.06.14 │ 2022.06.14 │ yes       │ 2022.06.1… │ 2022.06.14  │\n",
       "│ clinvar_a… │ ClinVar    │ annotator │ 14.3 MB  │ 2021.10.01 │ 2021.10.01 │ yes       │ 2021.10.01 │ 2021.10.01  │\n",
       "│            │ ACMG       │           │          │            │            │           │            │             │\n",
       "└────────────┴────────────┴───────────┴──────────┴────────────┴────────────┴───────────┴────────────┴─────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "_ = ov.module.ls(available=True, module_names=[\"clin.*\"], outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f387df9e-734f-42b9-b766-efb26e1c0028",
   "metadata": {},
   "source": [
    "_Get information on one module_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "be962aaf-786f-4341-92b3-555c4a5ca7b5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'readme': '# ClinVar: public archive of interpretations of clinically relevant variants\\nClinVar is a freely accessible, public archive of reports of the relationships among human variations and phenotypes, with supporting evidence. \\n \\nClinical Significance includes:  \\nBenign, Likely benign, Uncertain significance, Likely pathogenic, Pathogenic, Drug response, Association, Risk factor, Protective, Affects, Conflicting data from submitters, Other and Not provided\\n \\nReview Status is the level of review supporting the assertion of clinical significance for the variation. There are 8 different statuses to consider, ranked from highest to lowest levels of review:\\n\\n1. Practice Guideline\\n2. Reviewed by expert panel \\n3. Criteria provided, multiple submitters, no conflicts \\n4. Criteria provided, conflicting interpretations\\n5. Criteria provided, single submitter\\n6. No assertion for the individual variant \\n7. No assertion criteria provided\\n8. No assertion provided\\n\\nSee [Clinvar Review Status](https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/) for more information.\\n\\n![Screenshot](clinvar_screenshot_1.png)\\n<br />',\n",
       " 'store_availability': True,\n",
       " 'name': 'clinvar',\n",
       " 'title': 'ClinVar',\n",
       " 'description': 'ClinVar is an archive of reports of the relationships among human variations and phenotypes, with supporting evidence.',\n",
       " 'type': 'annotator',\n",
       " 'size': 45194349,\n",
       " 'tags': ['variants', 'clinical relevance', 'literature'],\n",
       " 'versions': {'2018.04.06': {'data_version': '2018.04.06',\n",
       "   'data_source': '2018.04.06',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2018.04.061': {'data_version': '2018.04.061',\n",
       "   'data_source': '2018.04.06',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2018.04.062': {'data_version': '2018.04.061',\n",
       "   'data_source': '2018.04.06',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.01.021': {'data_version': '2018.04.061',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.01.022': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.01.024': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.01.025': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.02.006': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.03.20': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.03.26': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.03.261': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.05.13': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.08.02': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2019.08.23': {'data_version': '2019.01.022',\n",
       "   'data_source': '2019.01.02',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2020.02.11': {'data_version': '2020.02.11',\n",
       "   'data_source': '2020.02.11',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2020.02.11.1': {'data_version': '2020.02.11',\n",
       "   'data_source': '2020.02.11',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2020.11.02': {'data_version': '2020.11.02',\n",
       "   'data_source': '2020.02.11',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.01.12': {'data_version': '2021.01.12',\n",
       "   'data_source': '2021.01.11',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.02.04': {'data_version': '2021.02.04',\n",
       "   'data_source': '2021.02.04',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.03.03': {'data_version': '2021.03.03',\n",
       "   'data_source': '2021.03.03',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.04.01': {'data_version': '2021.04.01',\n",
       "   'data_source': '2021.04.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.05.01': {'data_version': '2021.05.01',\n",
       "   'data_source': '2021.05.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.06.01': {'data_version': '2021.06.01',\n",
       "   'data_source': '2021.06.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.08.01': {'data_version': '2021.08.01',\n",
       "   'data_source': '2021.08.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.09.01': {'data_version': '2021.09.01',\n",
       "   'data_source': '2021.09.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2021.10.01': {'data_version': '2021.10.01',\n",
       "   'data_source': '2021.10.01',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2022.03.06': {'data_version': '2022.03.06',\n",
       "   'data_source': '2022.03.06',\n",
       "   'min_pkg_ver': '2.5.66'},\n",
       "  '2022.06.14': {'data_version': '2022.06.14',\n",
       "   'data_source': '2022.06.14',\n",
       "   'min_pkg_ver': '2.5.66'}},\n",
       " 'developer': {'module': {'name': 'NCBI',\n",
       "   'email': 'support@cravat.us',\n",
       "   'organization': 'NCBI',\n",
       "   'citation': 'Landrum, M.J., Lee, J.M., Benson, M., et al. (2016). ClinVar: public archive of interpretations of clinically relevant variants. Nucleic Acids Research, 44(D1):D862-8.',\n",
       "   'website': 'https://www.ncbi.nlm.nih.gov/clinvar/'},\n",
       "  'data': {'name': '',\n",
       "   'email': '',\n",
       "   'organization': '',\n",
       "   'citation': '',\n",
       "   'website': ''}},\n",
       " 'groups': [],\n",
       " 'output_columns': [{'name': 'sig',\n",
       "   'title': 'Clinical Significance',\n",
       "   'desc': ''},\n",
       "  {'name': 'disease_refs',\n",
       "   'title': 'Disease Ref Nums',\n",
       "   'desc': 'Disease reference numbers'},\n",
       "  {'name': 'disease_names', 'title': 'Disease Names', 'desc': ''},\n",
       "  {'name': 'rev_stat',\n",
       "   'title': 'Review Status',\n",
       "   'desc': 'The level of review supporting the assertion of clinical significance'},\n",
       "  {'name': 'id', 'title': 'ClinVar ID', 'desc': ''},\n",
       "  {'name': 'sig_conf',\n",
       "   'title': 'Significance Detail',\n",
       "   'desc': 'Addional detail on clinical significance used when it is conflicting'}],\n",
       " 'requires': ['wgclinvar'],\n",
       " 'min_pkg_ver': '2.5.66',\n",
       " 'installed': True,\n",
       " 'latest_installed': True,\n",
       " 'latest_store_version': '2022.06.14',\n",
       " 'latest_version': '2022.06.14.1'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.module.info(module_name=\"clinvar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ef794c-9f58-4aec-beb1-925fcdee2bae",
   "metadata": {},
   "source": [
    "_Give outer=ov.stdouter to print the module information to stdout._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b592d5ba-fac8-4cf9-b5f8-e8f8e4c99323",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">                                                      README                                                       </span>\n",
       "┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ # ClinVar: public archive of interpretations of clinically relevant variants                                    │\n",
       "│ ClinVar is a freely accessible, public archive of reports of the relationships among human variations and       │\n",
       "│ phenotypes, with supporting evidence.                                                                           │\n",
       "│                                                                                                                 │\n",
       "│ Clinical Significance includes:                                                                                 │\n",
       "│ Benign, Likely benign, Uncertain significance, Likely pathogenic, Pathogenic, Drug response, Association, Risk  │\n",
       "│ factor, Protective, Affects, Conflicting data from submitters, Other and Not provided                           │\n",
       "│                                                                                                                 │\n",
       "│ Review Status is the level of review supporting the assertion of clinical significance for the variation. There │\n",
       "│ are 8 different statuses to consider, ranked from highest to lowest levels of review:                           │\n",
       "│                                                                                                                 │\n",
       "│ 1. Practice Guideline                                                                                           │\n",
       "│ 2. Reviewed by expert panel                                                                                     │\n",
       "│ 3. Criteria provided, multiple submitters, no conflicts                                                         │\n",
       "│ 4. Criteria provided, conflicting interpretations                                                               │\n",
       "│ 5. Criteria provided, single submitter                                                                          │\n",
       "│ 6. No assertion for the individual variant                                                                      │\n",
       "│ 7. No assertion criteria provided                                                                               │\n",
       "│ 8. No assertion provided                                                                                        │\n",
       "│                                                                                                                 │\n",
       "│ See [Clinvar Review Status](https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/) for more information.     │\n",
       "│                                                                                                                 │\n",
       "│ ![Screenshot](clinvar_screenshot_1.png)                                                                         │\n",
       "│ &lt;br /&gt;                                                                                                          │\n",
       "└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m                                                      README                                                       \u001b[0m\n",
       "┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ # ClinVar: public archive of interpretations of clinically relevant variants                                    │\n",
       "│ ClinVar is a freely accessible, public archive of reports of the relationships among human variations and       │\n",
       "│ phenotypes, with supporting evidence.                                                                           │\n",
       "│                                                                                                                 │\n",
       "│ Clinical Significance includes:                                                                                 │\n",
       "│ Benign, Likely benign, Uncertain significance, Likely pathogenic, Pathogenic, Drug response, Association, Risk  │\n",
       "│ factor, Protective, Affects, Conflicting data from submitters, Other and Not provided                           │\n",
       "│                                                                                                                 │\n",
       "│ Review Status is the level of review supporting the assertion of clinical significance for the variation. There │\n",
       "│ are 8 different statuses to consider, ranked from highest to lowest levels of review:                           │\n",
       "│                                                                                                                 │\n",
       "│ 1. Practice Guideline                                                                                           │\n",
       "│ 2. Reviewed by expert panel                                                                                     │\n",
       "│ 3. Criteria provided, multiple submitters, no conflicts                                                         │\n",
       "│ 4. Criteria provided, conflicting interpretations                                                               │\n",
       "│ 5. Criteria provided, single submitter                                                                          │\n",
       "│ 6. No assertion for the individual variant                                                                      │\n",
       "│ 7. No assertion criteria provided                                                                               │\n",
       "│ 8. No assertion provided                                                                                        │\n",
       "│                                                                                                                 │\n",
       "│ See [Clinvar Review Status](https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/) for more information.     │\n",
       "│                                                                                                                 │\n",
       "│ ![Screenshot](clinvar_screenshot_1.png)                                                                         │\n",
       "│ <br />                                                                                                          │\n",
       "└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">┌──────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ store_availability   │ True                                                                                     │\n",
       "│ name                 │ clinvar                                                                                  │\n",
       "│ title                │ ClinVar                                                                                  │\n",
       "│ description          │ ClinVar is an archive of reports of the relationships among human variations and         │\n",
       "│                      │ phenotypes, with supporting evidence.                                                    │\n",
       "│ type                 │ annotator                                                                                │\n",
       "│ size                 │ 45194349                                                                                 │\n",
       "│ tags                 │ variants, clinical relevance, literature                                                 │\n",
       "│ groups               │                                                                                          │\n",
       "│ requires             │ wgclinvar                                                                                │\n",
       "│ min_pkg_ver          │ 2.5.66                                                                                   │\n",
       "│ installed            │ True                                                                                     │\n",
       "│ latest_installed     │ True                                                                                     │\n",
       "│ latest_store_version │ 2022.06.14                                                                               │\n",
       "│ latest_version       │ 2022.06.14.1                                                                             │\n",
       "└──────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "┌──────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ store_availability   │ True                                                                                     │\n",
       "│ name                 │ clinvar                                                                                  │\n",
       "│ title                │ ClinVar                                                                                  │\n",
       "│ description          │ ClinVar is an archive of reports of the relationships among human variations and         │\n",
       "│                      │ phenotypes, with supporting evidence.                                                    │\n",
       "│ type                 │ annotator                                                                                │\n",
       "│ size                 │ 45194349                                                                                 │\n",
       "│ tags                 │ variants, clinical relevance, literature                                                 │\n",
       "│ groups               │                                                                                          │\n",
       "│ requires             │ wgclinvar                                                                                │\n",
       "│ min_pkg_ver          │ 2.5.66                                                                                   │\n",
       "│ installed            │ True                                                                                     │\n",
       "│ latest_installed     │ True                                                                                     │\n",
       "│ latest_store_version │ 2022.06.14                                                                               │\n",
       "│ latest_version       │ 2022.06.14.1                                                                             │\n",
       "└──────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">                                                    Developers                                                     </span>\n",
       "┌──────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ <span style=\"font-weight: bold\">Module</span>       │                                                                                                  │\n",
       "│ Name         │ NCBI                                                                                             │\n",
       "│ Organization │ NCBI                                                                                             │\n",
       "│ Email        │ support@cravat.us                                                                                │\n",
       "│ Website      │ https://www.ncbi.nlm.nih.gov/clinvar/                                                            │\n",
       "│ Citation     │ Landrum, M.J., Lee, J.M., Benson, M., et al. (2016). ClinVar: public archive of interpretations  │\n",
       "│              │ of clinically relevant variants. Nucleic Acids Research, 44(D1):D862-8.                          │\n",
       "│ <span style=\"font-weight: bold\">Data</span>         │                                                                                                  │\n",
       "│ Name         │                                                                                                  │\n",
       "│ Organization │                                                                                                  │\n",
       "│ Email        │                                                                                                  │\n",
       "│ Website      │                                                                                                  │\n",
       "│ Citation     │                                                                                                  │\n",
       "└──────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m                                                    Developers                                                     \u001b[0m\n",
       "┌──────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────┐\n",
       "│ \u001b[1mModule\u001b[0m       │                                                                                                  │\n",
       "│ Name         │ NCBI                                                                                             │\n",
       "│ Organization │ NCBI                                                                                             │\n",
       "│ Email        │ support@cravat.us                                                                                │\n",
       "│ Website      │ https://www.ncbi.nlm.nih.gov/clinvar/                                                            │\n",
       "│ Citation     │ Landrum, M.J., Lee, J.M., Benson, M., et al. (2016). ClinVar: public archive of interpretations  │\n",
       "│              │ of clinically relevant variants. Nucleic Acids Research, 44(D1):D862-8.                          │\n",
       "│ \u001b[1mData\u001b[0m         │                                                                                                  │\n",
       "│ Name         │                                                                                                  │\n",
       "│ Organization │                                                                                                  │\n",
       "│ Email        │                                                                                                  │\n",
       "│ Website      │                                                                                                  │\n",
       "│ Citation     │                                                                                                  │\n",
       "└──────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">                               Versions                               </span>\n",
       "┌──────────────┬──────────────┬─────────────┬────────────────────────┐\n",
       "│<span style=\"font-weight: bold\"> Version      </span>│<span style=\"font-weight: bold\"> Data version </span>│<span style=\"font-weight: bold\"> Data source </span>│<span style=\"font-weight: bold\"> Minimum OakVar version </span>│\n",
       "├──────────────┼──────────────┼─────────────┼────────────────────────┤\n",
       "│ 2018.04.06   │ 2018.04.06   │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2018.04.061  │ 2018.04.061  │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2018.04.062  │ 2018.04.061  │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2019.01.021  │ 2018.04.061  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.022  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.024  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.025  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.02.006  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.20   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.26   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.261  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.05.13   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.08.02   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.08.23   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2020.02.11   │ 2020.02.11   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2020.02.11.1 │ 2020.02.11   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2020.11.02   │ 2020.11.02   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2021.01.12   │ 2021.01.12   │ 2021.01.11  │ 2.5.66                 │\n",
       "│ 2021.02.04   │ 2021.02.04   │ 2021.02.04  │ 2.5.66                 │\n",
       "│ 2021.03.03   │ 2021.03.03   │ 2021.03.03  │ 2.5.66                 │\n",
       "│ 2021.04.01   │ 2021.04.01   │ 2021.04.01  │ 2.5.66                 │\n",
       "│ 2021.05.01   │ 2021.05.01   │ 2021.05.01  │ 2.5.66                 │\n",
       "│ 2021.06.01   │ 2021.06.01   │ 2021.06.01  │ 2.5.66                 │\n",
       "│ 2021.08.01   │ 2021.08.01   │ 2021.08.01  │ 2.5.66                 │\n",
       "│ 2021.09.01   │ 2021.09.01   │ 2021.09.01  │ 2.5.66                 │\n",
       "│ 2021.10.01   │ 2021.10.01   │ 2021.10.01  │ 2.5.66                 │\n",
       "│ 2022.03.06   │ 2022.03.06   │ 2022.03.06  │ 2.5.66                 │\n",
       "│ 2022.06.14   │ 2022.06.14   │ 2022.06.14  │ 2.5.66                 │\n",
       "└──────────────┴──────────────┴─────────────┴────────────────────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m                               Versions                               \u001b[0m\n",
       "┌──────────────┬──────────────┬─────────────┬────────────────────────┐\n",
       "│\u001b[1m \u001b[0m\u001b[1mVersion     \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mData version\u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mData source\u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mMinimum OakVar version\u001b[0m\u001b[1m \u001b[0m│\n",
       "├──────────────┼──────────────┼─────────────┼────────────────────────┤\n",
       "│ 2018.04.06   │ 2018.04.06   │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2018.04.061  │ 2018.04.061  │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2018.04.062  │ 2018.04.061  │ 2018.04.06  │ 2.5.66                 │\n",
       "│ 2019.01.021  │ 2018.04.061  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.022  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.024  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.01.025  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.02.006  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.20   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.26   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.03.261  │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.05.13   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.08.02   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2019.08.23   │ 2019.01.022  │ 2019.01.02  │ 2.5.66                 │\n",
       "│ 2020.02.11   │ 2020.02.11   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2020.02.11.1 │ 2020.02.11   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2020.11.02   │ 2020.11.02   │ 2020.02.11  │ 2.5.66                 │\n",
       "│ 2021.01.12   │ 2021.01.12   │ 2021.01.11  │ 2.5.66                 │\n",
       "│ 2021.02.04   │ 2021.02.04   │ 2021.02.04  │ 2.5.66                 │\n",
       "│ 2021.03.03   │ 2021.03.03   │ 2021.03.03  │ 2.5.66                 │\n",
       "│ 2021.04.01   │ 2021.04.01   │ 2021.04.01  │ 2.5.66                 │\n",
       "│ 2021.05.01   │ 2021.05.01   │ 2021.05.01  │ 2.5.66                 │\n",
       "│ 2021.06.01   │ 2021.06.01   │ 2021.06.01  │ 2.5.66                 │\n",
       "│ 2021.08.01   │ 2021.08.01   │ 2021.08.01  │ 2.5.66                 │\n",
       "│ 2021.09.01   │ 2021.09.01   │ 2021.09.01  │ 2.5.66                 │\n",
       "│ 2021.10.01   │ 2021.10.01   │ 2021.10.01  │ 2.5.66                 │\n",
       "│ 2022.03.06   │ 2022.03.06   │ 2022.03.06  │ 2.5.66                 │\n",
       "│ 2022.06.14   │ 2022.06.14   │ 2022.06.14  │ 2.5.66                 │\n",
       "└──────────────┴──────────────┴─────────────┴────────────────────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">                                                      Output                                                       </span>\n",
       "┌───────────────┬───────────────────────┬────────────────────────────────────────────────────────────────┬────────┐\n",
       "│<span style=\"font-weight: bold\"> Name          </span>│<span style=\"font-weight: bold\"> Title                 </span>│<span style=\"font-weight: bold\"> Description                                                    </span>│<span style=\"font-weight: bold\"> Type   </span>│\n",
       "├───────────────┼───────────────────────┼────────────────────────────────────────────────────────────────┼────────┤\n",
       "│ sig           │ Clinical Significance │                                                                │ string │\n",
       "│ disease_refs  │ Disease Ref Nums      │ Disease reference numbers                                      │ string │\n",
       "│ disease_names │ Disease Names         │                                                                │ string │\n",
       "│ rev_stat      │ Review Status         │ The level of review supporting the assertion of clinical       │ string │\n",
       "│               │                       │ significance                                                   │        │\n",
       "│ id            │ ClinVar ID            │                                                                │ string │\n",
       "│ sig_conf      │ Significance Detail   │ Addional detail on clinical significance used when it is       │ string │\n",
       "│               │                       │ conflicting                                                    │        │\n",
       "└───────────────┴───────────────────────┴────────────────────────────────────────────────────────────────┴────────┘\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m                                                      Output                                                       \u001b[0m\n",
       "┌───────────────┬───────────────────────┬────────────────────────────────────────────────────────────────┬────────┐\n",
       "│\u001b[1m \u001b[0m\u001b[1mName         \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mTitle                \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mDescription                                                   \u001b[0m\u001b[1m \u001b[0m│\u001b[1m \u001b[0m\u001b[1mType  \u001b[0m\u001b[1m \u001b[0m│\n",
       "├───────────────┼───────────────────────┼────────────────────────────────────────────────────────────────┼────────┤\n",
       "│ sig           │ Clinical Significance │                                                                │ string │\n",
       "│ disease_refs  │ Disease Ref Nums      │ Disease reference numbers                                      │ string │\n",
       "│ disease_names │ Disease Names         │                                                                │ string │\n",
       "│ rev_stat      │ Review Status         │ The level of review supporting the assertion of clinical       │ string │\n",
       "│               │                       │ significance                                                   │        │\n",
       "│ id            │ ClinVar ID            │                                                                │ string │\n",
       "│ sig_conf      │ Significance Detail   │ Addional detail on clinical significance used when it is       │ string │\n",
       "│               │                       │ conflicting                                                    │        │\n",
       "└───────────────┴───────────────────────┴────────────────────────────────────────────────────────────────┴────────┘\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "_ = ov.module.info(module_name=\"clinvar\", outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "827e036b-9941-4785-94ac-561bae06c1ca",
   "metadata": {},
   "source": [
    "_Install modules_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6ca2667f-9fb0-47a5-a921-92e99ef9af82",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No store update to fetch\n",
      "The following modules will be installed:\n",
      "- biogrid==3.7.1\n",
      "- tsvreporter==2.1.0\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Proceed? ([y]/n) >  y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Installing biogrid...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "biogrid==3.7.1 already exists.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Installing tsvreporter...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "tsvreporter: Local version (2.1.1) is higher than the latest store version (2.1.0). Use --overwrite to overwrite.\n"
     ]
    }
   ],
   "source": [
    "ov.module.install(module_names=[\"biogrid\", \"tsvreporter\"], outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "049b3f8b-52d3-4843-8d17-f0ef8838c1c9",
   "metadata": {},
   "source": [
    "_Uninstall modules_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "229766b6-c90e-4679-832a-06f6ca91b9be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uninstalling:\n",
      "- biogrid\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Proceed? (y/N) >  y\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.module.uninstall(module_names=[\"biogrid\"], outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4faae67f-6f26-4865-b5a3-3346de2cc7b6",
   "metadata": {},
   "source": [
    "_Let's bring back the biogrid module back._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dcc33f20-669a-48ff-b60c-523ce9123b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "ov.module.install(module_names=[\"biogrid\"], yes=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a07d2f47-def7-4b59-a9e8-6bf22285edfe",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Annotate and reports"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7a5ad88-f0f4-4b0e-af22-f9cddc426b24",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "##### Create an example input file in the current directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f520389-9fe5-4a5a-9a05-b99f0584695b",
   "metadata": {},
   "source": [
    "> This should create a file `exampleinput` in the current directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2ed6d4bb-184f-4101-9431-2dfca4fc0758",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('exampleinput')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.api.new.exampleinput()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba74e028-d8a4-4594-bda0-653913a14bb4",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Annotate the example input file with the default gene mapper."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2db534e8-1a1c-4802-8188-2af7022325fc",
   "metadata": {},
   "source": [
    "> This should have created a database file `exampleinput.sqlite`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d20dc2fa-f29d-4ba0-8ec5-831b2577eae9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ov.api.run(inputs=[\"exampleinput\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c952a349-8614-4610-af04-f3d5048bfb82",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### Annotation data in the database file can be loaded into Polars dataframe."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c725d27-b496-4ba1-b785-6137c1508576",
   "metadata": {},
   "source": [
    "> Variant level (default)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "989e7f50-6fd8-4ca9-ab90-f4e069b2bd45",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ov.get_df_from_db(\"exampleinput.sqlite\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cb08613-6bb8-4e21-a90c-5c89d6e58ac3",
   "metadata": {},
   "source": [
    "> Gene level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa4efbc-5188-4f98-91f8-e75ba341ff6a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ov.get_df_from_db(\"exampleinput.sqlite\", \"gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40241dec-2b6c-44dc-be5c-3a738f79cbb8",
   "metadata": {},
   "source": [
    "##### Annotate with ClinVar and BioGRID annotators. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3347e088-db8d-4197-9c9b-f74b6c001876",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023/01/30 21:44:09 oakvar               /home/rick/mambaforge/envs/ov/lib/python3.8/site-packages/ipykernel_launcher.py -f /home/rick/.local/share/jupyter/runtime/kernel-09c2490f-cd07-4709-88e2-ee822f2fe969.json\n",
      "2023/01/30 21:44:09 oakvar               started: Mon Jan 30 21:44:09 2023\n",
      "2023/01/30 21:44:09 oakvar               conf file: None\n",
      "2023/01/30 21:44:09 oakvar               started OakVar\n",
      "2023/01/30 21:44:09 oakvar               input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:44:09 oakvar               system: oakvar==2.7.35 /mnt/d/Git/oakvar/oakvar\n",
      "2023/01/30 21:44:09 oakvar               module: clinvar==2022.06.14.1 /mnt/e/oakvar/annotators/clinvar/clinvar.py\n",
      "2023/01/30 21:44:09 oakvar               module: biogrid==3.7.1 /mnt/e/oakvar/annotators/biogrid/biogrid.py\n",
      "2023/01/30 21:44:09 oakvar               module: gencode==41.3.0 /mnt/e/oakvar/mappers/gencode/gencode.py\n",
      "2023/01/30 21:44:09 oakvar               starting converter step...\n",
      "2023/01/30 21:44:09 oakvar               running converter\n",
      "2023/01/30 21:44:09 oakvar               master_converter. path=/mnt/d/Git/oakvar/oakvar/lib/base/master_converter.py\n",
      "2023/01/30 21:44:09 oakvar               master_converter. inputs=['/mnt/d/Git/oakvar/extras/notebooks/exampleinput']\n",
      "2023/01/30 21:44:09 oakvar               master_converter. name=exampleinput\n",
      "2023/01/30 21:44:09 oakvar               master_converter. output_dir=/mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:44:09 oakvar               master_converter. input_encoding=None\n",
      "2023/01/30 21:44:09 oakvar               master_converter. conf={}\n",
      "2023/01/30 21:44:10 oakvar.converter     started converter\n",
      "2023/01/30 21:44:10 oakvar.converter     detecting encoding of /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:44:10 oakvar.converter     encoding: /mnt/d/Git/oakvar/extras/notebooks/exampleinput utf-8\n",
      "2023/01/30 21:44:10 oakvar.converter     using cravat for /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:44:10 oakvar.converter     module: cravat-converter==1.1.2 /mnt/e/oakvar/converters/cravat-converter/cravat-converter.py\n",
      "2023/01/30 21:44:10 oakvar.converter     input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:44:10 oakvar.converter     input format: cravat\n",
      "2023/01/30 21:44:10 oakvar.converter     genome_assembly: hg38\n",
      "2023/01/30 21:44:10 oakvar.converter     liftover needed: False\n",
      "2023/01/30 21:44:10 oakvar.converter     liftover for chrM needed: False\n",
      "2023/01/30 21:44:10 oakvar.converter     number of valid variants: 373\n",
      "2023/01/30 21:44:10 oakvar.converter     number of error lines: 0\n",
      "2023/01/30 21:44:10 oakvar.converter     total number of converted variants: 373\n",
      "2023/01/30 21:44:10 oakvar.converter     number of total error lines: 0\n",
      "2023/01/30 21:44:10 oakvar.converter     finished: Mon Jan 30 21:44:10 2023\n",
      "2023/01/30 21:44:10 oakvar.converter     runtime: 0.027\n",
      "2023/01/30 21:44:10 oakvar.converter     finished Converter\n",
      "2023/01/30 21:44:10 oakvar               converter step finished in 0.000s\n",
      "2023/01/30 21:44:10 oakvar               starting preparer step...\n",
      "2023/01/30 21:44:10 oakvar               preparer step finished in 0.000s\n",
      "2023/01/30 21:44:10 oakvar               starting mapper step...\n",
      "2023/01/30 21:44:10 oakvar               num_workers: 1\n",
      "2023/01/30 21:44:10 oakvar               input line chunksize=373 total number of input lines=382 number of chunks=1\n",
      "2023/01/30 21:44:10 oakvar.mapper        input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crv\n",
      "2023/01/30 21:44:12 oakvar.mapper        started: Mon Jan 30 21:44:12 2023 | 0\n",
      "2023/01/30 21:44:12 oakvar.mapper        started GENCODE Mapper (gencode)\n",
      "2023/01/30 21:44:13 oakvar.mapper        finished: Mon Jan 30 21:44:13 2023 | 0\n",
      "2023/01/30 21:44:13 oakvar.mapper        runtime:  0.169\n",
      "2023/01/30 21:44:13 oakvar               mapper step finished in 0.000s\n",
      "2023/01/30 21:44:13 oakvar               starting annotator step...\n",
      "2023/01/30 21:44:13 oakvar               num_workers: 1\n",
      "2023/01/30 21:44:13 oakvar.clinvar       started ClinVar (clinvar)\n",
      "2023/01/30 21:44:13 oakvar.clinvar       started: Mon Jan 30 21:44:13 2023\n",
      "2023/01/30 21:44:13 oakvar.clinvar       started ClinVar (clinvar)\n",
      "2023/01/30 21:44:13 oakvar.clinvar       clinvar: finished at Mon Jan 30 21:44:13 2023\n",
      "2023/01/30 21:44:13 oakvar.clinvar       clinvar: runtime 0.068s\n",
      "2023/01/30 21:44:14 oakvar.biogrid       started BioGRID (biogrid)\n",
      "2023/01/30 21:44:14 oakvar.biogrid       started: Mon Jan 30 21:44:14 2023\n",
      "2023/01/30 21:44:14 oakvar.biogrid       started BioGRID (biogrid)\n",
      "2023/01/30 21:44:14 oakvar.biogrid       biogrid: finished at Mon Jan 30 21:44:14 2023\n",
      "2023/01/30 21:44:14 oakvar.biogrid       biogrid: runtime 0.349s\n",
      "2023/01/30 21:44:15 oakvar               annotator step finished in 0.000s\n",
      "2023/01/30 21:44:15 oakvar               starting aggregator step...\n",
      "2023/01/30 21:44:15 oakvar               running Aggregator (variant)\n",
      "2023/01/30 21:44:15 oakvar.aggregator    level: variant\n",
      "2023/01/30 21:44:15 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:44:16 oakvar.aggregator    started aggregator (variant)\n",
      "2023/01/30 21:44:16 oakvar.aggregator    started: Mon Jan 30 21:44:16 2023\n",
      "2023/01/30 21:44:16 oakvar.aggregator    finished: Mon Jan 30 21:44:16 2023\n",
      "2023/01/30 21:44:16 oakvar.aggregator    runtime: 0.245\n",
      "2023/01/30 21:44:16 oakvar.aggregator    finished aggregator (variant)\n",
      "2023/01/30 21:44:16 oakvar               Aggregator variant finished in 0.000s\n",
      "2023/01/30 21:44:16 oakvar               running Aggregator (gene)\n",
      "2023/01/30 21:44:16 oakvar.aggregator    level: gene\n",
      "2023/01/30 21:44:16 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:44:17 oakvar.aggregator    started aggregator (gene)\n",
      "2023/01/30 21:44:17 oakvar.aggregator    started: Mon Jan 30 21:44:17 2023\n",
      "2023/01/30 21:44:17 oakvar.aggregator    finished: Mon Jan 30 21:44:17 2023\n",
      "2023/01/30 21:44:17 oakvar.aggregator    runtime: 0.332\n",
      "2023/01/30 21:44:18 oakvar.aggregator    finished aggregator (gene)\n",
      "2023/01/30 21:44:18 oakvar               Aggregator gene finished in 0.000s\n",
      "2023/01/30 21:44:18 oakvar               running Aggregator (sample)\n",
      "2023/01/30 21:44:18 oakvar.aggregator    level: sample\n",
      "2023/01/30 21:44:18 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:44:18 oakvar.aggregator    started aggregator (sample)\n",
      "2023/01/30 21:44:18 oakvar.aggregator    started: Mon Jan 30 21:44:18 2023\n",
      "2023/01/30 21:44:19 oakvar.aggregator    finished: Mon Jan 30 21:44:19 2023\n",
      "2023/01/30 21:44:19 oakvar.aggregator    runtime: 0.242\n",
      "2023/01/30 21:44:19 oakvar.aggregator    finished aggregator (sample)\n",
      "2023/01/30 21:44:19 oakvar               Aggregator sample finished in 0.000s\n",
      "2023/01/30 21:44:19 oakvar               running Aggregator (mapping)\n",
      "2023/01/30 21:44:19 oakvar.aggregator    level: mapping\n",
      "2023/01/30 21:44:19 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:44:19 oakvar.aggregator    started aggregator (mapping)\n",
      "2023/01/30 21:44:19 oakvar.aggregator    started: Mon Jan 30 21:44:19 2023\n",
      "2023/01/30 21:44:20 oakvar.aggregator    finished: Mon Jan 30 21:44:20 2023\n",
      "2023/01/30 21:44:20 oakvar.aggregator    runtime: 0.231\n",
      "2023/01/30 21:44:20 oakvar.aggregator    finished aggregator (mapping)\n",
      "2023/01/30 21:44:20 oakvar               Aggregator mapping finished in 0.000s\n",
      "2023/01/30 21:44:20 oakvar               aggregator step finished in 0.000s\n",
      "2023/01/30 21:44:20 oakvar               starting postaggregator step...\n",
      "2023/01/30 21:44:21 oakvar.tagsampler    started Samples and Tags (tagsampler)\n",
      "2023/01/30 21:44:21 oakvar.tagsampler    started: Mon Jan 30 21:44:21 2023\n",
      "2023/01/30 21:44:21 oakvar.tagsampler    finished: Mon Jan 30 21:44:21 2023\n",
      "2023/01/30 21:44:21 oakvar.tagsampler    runtime: 0.763\n",
      "2023/01/30 21:44:21 oakvar.tagsampler    Finished Samples and Tags (tagsampler)\n",
      "2023/01/30 21:44:21 oakvar               tagsampler finished in 0.000s\n",
      "2023/01/30 21:44:22 oakvar               vcfinfo finished in 0.000s\n",
      "2023/01/30 21:44:22 oakvar               postaggregator step finished in 0.000s\n",
      "2023/01/30 21:44:22 oakvar               finished: Mon Jan 30 21:44:22 2023\n",
      "2023/01/30 21:44:22 oakvar               finished normally. Runtime: 12.070s\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.biogrid.gen\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.clinvar.var\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crg\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crm\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crs\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crv\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crx\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.err\n",
      "2023/01/30 21:44:22 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.original_input.var\n"
     ]
    }
   ],
   "source": [
    "ov.run(inputs=[\"exampleinput\"], annotators=[\"clinvar\", \"biogrid\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "917db383-c86e-4229-ac05-7e26a7ffc40d",
   "metadata": {},
   "source": [
    "##### Generate a VCF file of the annotated variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7241b6f4-f102-453d-919e-2bd6aceb167c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating vcf report...\n",
      "report created: exampleinput.vcf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'vcf': 'exampleinput.vcf'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.report(dbpath=\"exampleinput.sqlite\", report_types=[\"vcf\"], outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3611e618-2a25-4d43-852e-869e02e695a6",
   "metadata": {},
   "source": [
    "##### Annotation and report generation can be done at once. Also, reports of different formats can be generated in one go."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0294224b-a62b-4ff6-bd08-be7e2af26ac8",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023/01/30 21:51:31 oakvar               /home/rick/mambaforge/envs/ov/lib/python3.8/site-packages/ipykernel_launcher.py -f /home/rick/.local/share/jupyter/runtime/kernel-09c2490f-cd07-4709-88e2-ee822f2fe969.json\n",
      "2023/01/30 21:51:31 oakvar               started: Mon Jan 30 21:51:31 2023\n",
      "2023/01/30 21:51:31 oakvar               conf file: None\n",
      "2023/01/30 21:51:31 oakvar               started OakVar\n",
      "2023/01/30 21:51:31 oakvar               input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:51:31 oakvar               system: oakvar==2.7.35 /mnt/d/Git/oakvar/oakvar\n",
      "2023/01/30 21:51:31 oakvar               module: clinvar==2022.06.14.1 /mnt/e/oakvar/annotators/clinvar/clinvar.py\n",
      "2023/01/30 21:51:31 oakvar               module: biogrid==3.7.1 /mnt/e/oakvar/annotators/biogrid/biogrid.py\n",
      "2023/01/30 21:51:31 oakvar               module: gencode==41.3.0 /mnt/e/oakvar/mappers/gencode/gencode.py\n",
      "2023/01/30 21:51:31 oakvar               module: vcfreporter==4.3.0 /mnt/e/oakvar/reporters/vcfreporter/vcfreporter.py\n",
      "2023/01/30 21:51:31 oakvar               module: csvreporter==2.1.0 /mnt/e/oakvar/reporters/csvreporter/csvreporter.py\n",
      "2023/01/30 21:51:31 oakvar               starting converter step...\n",
      "2023/01/30 21:51:31 oakvar               running converter\n",
      "2023/01/30 21:51:31 oakvar               master_converter. path=/mnt/d/Git/oakvar/oakvar/lib/base/master_converter.py\n",
      "2023/01/30 21:51:31 oakvar               master_converter. inputs=['/mnt/d/Git/oakvar/extras/notebooks/exampleinput']\n",
      "2023/01/30 21:51:31 oakvar               master_converter. name=exampleinput\n",
      "2023/01/30 21:51:31 oakvar               master_converter. output_dir=/mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:51:31 oakvar               master_converter. input_encoding=None\n",
      "2023/01/30 21:51:31 oakvar               master_converter. conf={}\n",
      "2023/01/30 21:51:31 oakvar.converter     started converter\n",
      "2023/01/30 21:51:31 oakvar.converter     detecting encoding of /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:51:31 oakvar.converter     encoding: /mnt/d/Git/oakvar/extras/notebooks/exampleinput utf-8\n",
      "2023/01/30 21:51:31 oakvar.converter     using cravat for /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:51:31 oakvar.converter     module: cravat-converter==1.1.2 /mnt/e/oakvar/converters/cravat-converter/cravat-converter.py\n",
      "2023/01/30 21:51:31 oakvar.converter     input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput\n",
      "2023/01/30 21:51:31 oakvar.converter     input format: cravat\n",
      "2023/01/30 21:51:31 oakvar.converter     genome_assembly: hg38\n",
      "2023/01/30 21:51:31 oakvar.converter     liftover needed: False\n",
      "2023/01/30 21:51:31 oakvar.converter     liftover for chrM needed: False\n",
      "2023/01/30 21:51:31 oakvar.converter     number of valid variants: 373\n",
      "2023/01/30 21:51:31 oakvar.converter     number of error lines: 0\n",
      "2023/01/30 21:51:31 oakvar.converter     total number of converted variants: 373\n",
      "2023/01/30 21:51:31 oakvar.converter     number of total error lines: 0\n",
      "2023/01/30 21:51:31 oakvar.converter     finished: Mon Jan 30 21:51:31 2023\n",
      "2023/01/30 21:51:31 oakvar.converter     runtime: 0.029\n",
      "2023/01/30 21:51:31 oakvar.converter     finished Converter\n",
      "2023/01/30 21:51:31 oakvar               converter step finished in 0.000s\n",
      "2023/01/30 21:51:31 oakvar               starting preparer step...\n",
      "2023/01/30 21:51:31 oakvar               preparer step finished in 0.000s\n",
      "2023/01/30 21:51:31 oakvar               starting mapper step...\n",
      "2023/01/30 21:51:31 oakvar               num_workers: 1\n",
      "2023/01/30 21:51:31 oakvar               input line chunksize=373 total number of input lines=382 number of chunks=1\n",
      "2023/01/30 21:51:31 oakvar.mapper        input file: /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crv\n",
      "2023/01/30 21:51:33 oakvar.mapper        started: Mon Jan 30 21:51:33 2023 | 0\n",
      "2023/01/30 21:51:33 oakvar.mapper        started GENCODE Mapper (gencode)\n",
      "2023/01/30 21:51:33 oakvar.mapper        finished: Mon Jan 30 21:51:33 2023 | 0\n",
      "2023/01/30 21:51:33 oakvar.mapper        runtime:  0.170\n",
      "2023/01/30 21:51:33 oakvar               mapper step finished in 0.000s\n",
      "2023/01/30 21:51:33 oakvar               starting annotator step...\n",
      "2023/01/30 21:51:33 oakvar               num_workers: 1\n",
      "2023/01/30 21:51:33 oakvar.clinvar       started ClinVar (clinvar)\n",
      "2023/01/30 21:51:33 oakvar.clinvar       started: Mon Jan 30 21:51:33 2023\n",
      "2023/01/30 21:51:33 oakvar.clinvar       started ClinVar (clinvar)\n",
      "2023/01/30 21:51:33 oakvar.clinvar       clinvar: finished at Mon Jan 30 21:51:33 2023\n",
      "2023/01/30 21:51:33 oakvar.clinvar       clinvar: runtime 0.076s\n",
      "2023/01/30 21:51:33 oakvar.biogrid       started BioGRID (biogrid)\n",
      "2023/01/30 21:51:33 oakvar.biogrid       started: Mon Jan 30 21:51:33 2023\n",
      "2023/01/30 21:51:34 oakvar.biogrid       started BioGRID (biogrid)\n",
      "2023/01/30 21:51:34 oakvar.biogrid       biogrid: finished at Mon Jan 30 21:51:34 2023\n",
      "2023/01/30 21:51:34 oakvar.biogrid       biogrid: runtime 0.329s\n",
      "2023/01/30 21:51:35 oakvar               annotator step finished in 0.000s\n",
      "2023/01/30 21:51:35 oakvar               starting aggregator step...\n",
      "2023/01/30 21:51:35 oakvar               running Aggregator (variant)\n",
      "2023/01/30 21:51:35 oakvar.aggregator    level: variant\n",
      "2023/01/30 21:51:35 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:51:35 oakvar.aggregator    started aggregator (variant)\n",
      "2023/01/30 21:51:35 oakvar.aggregator    started: Mon Jan 30 21:51:35 2023\n",
      "2023/01/30 21:51:35 oakvar.aggregator    finished: Mon Jan 30 21:51:35 2023\n",
      "2023/01/30 21:51:35 oakvar.aggregator    runtime: 0.225\n",
      "2023/01/30 21:51:36 oakvar.aggregator    finished aggregator (variant)\n",
      "2023/01/30 21:51:36 oakvar               Aggregator variant finished in 0.000s\n",
      "2023/01/30 21:51:36 oakvar               running Aggregator (gene)\n",
      "2023/01/30 21:51:36 oakvar.aggregator    level: gene\n",
      "2023/01/30 21:51:36 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:51:37 oakvar.aggregator    started aggregator (gene)\n",
      "2023/01/30 21:51:37 oakvar.aggregator    started: Mon Jan 30 21:51:37 2023\n",
      "2023/01/30 21:51:37 oakvar.aggregator    finished: Mon Jan 30 21:51:37 2023\n",
      "2023/01/30 21:51:37 oakvar.aggregator    runtime: 0.263\n",
      "2023/01/30 21:51:38 oakvar.aggregator    finished aggregator (gene)\n",
      "2023/01/30 21:51:38 oakvar               Aggregator gene finished in 0.000s\n",
      "2023/01/30 21:51:38 oakvar               running Aggregator (sample)\n",
      "2023/01/30 21:51:38 oakvar.aggregator    level: sample\n",
      "2023/01/30 21:51:38 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:51:38 oakvar.aggregator    started aggregator (sample)\n",
      "2023/01/30 21:51:38 oakvar.aggregator    started: Mon Jan 30 21:51:38 2023\n",
      "2023/01/30 21:51:39 oakvar.aggregator    finished: Mon Jan 30 21:51:39 2023\n",
      "2023/01/30 21:51:39 oakvar.aggregator    runtime: 0.224\n",
      "2023/01/30 21:51:39 oakvar.aggregator    finished aggregator (sample)\n",
      "2023/01/30 21:51:39 oakvar               Aggregator sample finished in 0.000s\n",
      "2023/01/30 21:51:39 oakvar               running Aggregator (mapping)\n",
      "2023/01/30 21:51:39 oakvar.aggregator    level: mapping\n",
      "2023/01/30 21:51:39 oakvar.aggregator    input directory: /mnt/d/Git/oakvar/extras/notebooks\n",
      "2023/01/30 21:51:39 oakvar.aggregator    started aggregator (mapping)\n",
      "2023/01/30 21:51:39 oakvar.aggregator    started: Mon Jan 30 21:51:39 2023\n",
      "2023/01/30 21:51:40 oakvar.aggregator    finished: Mon Jan 30 21:51:40 2023\n",
      "2023/01/30 21:51:40 oakvar.aggregator    runtime: 0.236\n",
      "2023/01/30 21:51:40 oakvar.aggregator    finished aggregator (mapping)\n",
      "2023/01/30 21:51:40 oakvar               Aggregator mapping finished in 0.000s\n",
      "2023/01/30 21:51:40 oakvar               aggregator step finished in 0.000s\n",
      "2023/01/30 21:51:40 oakvar               starting postaggregator step...\n",
      "2023/01/30 21:51:40 oakvar.tagsampler    started Samples and Tags (tagsampler)\n",
      "2023/01/30 21:51:40 oakvar.tagsampler    started: Mon Jan 30 21:51:40 2023\n",
      "2023/01/30 21:51:41 oakvar.tagsampler    finished: Mon Jan 30 21:51:41 2023\n",
      "2023/01/30 21:51:41 oakvar.tagsampler    runtime: 0.747\n",
      "2023/01/30 21:51:41 oakvar.tagsampler    Finished Samples and Tags (tagsampler)\n",
      "2023/01/30 21:51:41 oakvar               tagsampler finished in 0.000s\n",
      "2023/01/30 21:51:41 oakvar               vcfinfo finished in 0.000s\n",
      "2023/01/30 21:51:41 oakvar               postaggregator step finished in 0.000s\n",
      "2023/01/30 21:51:41 oakvar               starting reporter step...\n",
      "2023/01/30 21:51:41 oakvar               starting vcfreporter...\n",
      "2023/01/30 21:51:44 oakvar               vcfreporter finished in 0.000s\n",
      "2023/01/30 21:51:44 oakvar               report created: /mnt/d/Git/oakvar/extras/notebooks/exampleinput.vcf \n",
      "2023/01/30 21:51:44 oakvar               starting csvreporter...\n",
      "2023/01/30 21:51:46 oakvar               csvreporter finished in 0.000s\n",
      "2023/01/30 21:51:46 oakvar               report created: /mnt/d/Git/oakvar/extras/notebooks/exampleinput.variant.csv \n",
      "2023/01/30 21:51:46 oakvar               reporter step finished in 0.000s\n",
      "2023/01/30 21:51:46 oakvar               finished: Mon Jan 30 21:51:46 2023\n",
      "2023/01/30 21:51:46 oakvar               finished normally. Runtime: 15.508s\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.biogrid.gen\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.clinvar.var\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crg\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crm\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crs\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crv\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.crx\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.err\n",
      "2023/01/30 21:51:46 oakvar               removing /mnt/d/Git/oakvar/extras/notebooks/exampleinput.original_input.var\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'vcf': '/mnt/d/Git/oakvar/extras/notebooks/exampleinput.vcf',\n",
       " 'csv': ['/mnt/d/Git/oakvar/extras/notebooks/exampleinput.variant.csv']}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ov.run(inputs=[\"exampleinput\"], annotators=[\"clinvar\", \"biogrid\"], report_types=[\"vcf\", \"csv\"], outer=ov.stdouter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ee4d303-863f-487c-adbf-1593384e2ea6",
   "metadata": {},
   "source": [
    "## Module development"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fcbee32-dde2-43b7-91d2-2c761298cab5",
   "metadata": {},
   "source": [
    "### Annotator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5fe01771-c222-4c4d-b8e6-100adb179bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "import oakvar as ov\n",
    "from oakvar import BaseAnnotator\n",
    "class Annotator(BaseAnnotator):\n",
    "    def annotate_df(self, df):\n",
    "        return df\n",
    "    def annotate(self, input_data: dict):\n",
    "        out_data = input_data\n",
    "        out_data[\"col1\"] = \"test_output\"\n",
    "        return out_data\n",
    "m = Annotator(name=\"awesome\", level=\"variant\", output_columns=[{\"name\": \"col1\", \"type\": \"string\", \"title\": \"Col1\"}])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5510abb5-9348-49bb-a390-ca906b2f4952",
   "metadata": {},
   "outputs": [],
   "source": [
    "m = Annotator(name=\"awesome\", title=\"Awesome annotator\", level=\"variant\", input_file=\"exampleinput.crv\", \n",
    "              output_columns=[{\"name\": \"col1\", \"type\": \"string\", \"title\": \"Col1\"}],\n",
    "              input_columns=[\"chrom\", \"pos\", \"ref_base\", \"alt_base\", \"note\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a97637eb-d802-4b62-8d40-3be170089a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "m.save(overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d72aec1-331c-4349-b01d-15c7ae508b93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        white-space: pre;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-top: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-bottom: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        line-height: 95%;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\" >\n",
       "<small>shape: (372, 6)</small>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>\n",
       "uid\n",
       "</th>\n",
       "<th>\n",
       "chrom\n",
       "</th>\n",
       "<th>\n",
       "pos\n",
       "</th>\n",
       "<th>\n",
       "pos_end\n",
       "</th>\n",
       "<th>\n",
       "ref_base\n",
       "</th>\n",
       "<th>\n",
       "alt_base\n",
       "</th>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>\n",
       "2\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "2987654\n",
       "</td>\n",
       "<td>\n",
       "2987654\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "3\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "43077259\n",
       "</td>\n",
       "<td>\n",
       "43077259\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "4\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "8055656\n",
       "</td>\n",
       "<td>\n",
       "8055656\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "5\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864470\n",
       "</td>\n",
       "<td>\n",
       "87864470\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "6\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "7\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "87864487\n",
       "</td>\n",
       "<td>\n",
       "&quot;AA&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "8\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;CG&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "9\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;CT&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "10\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "100719861\n",
       "</td>\n",
       "<td>\n",
       "100719861\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "11\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "10100\n",
       "</td>\n",
       "<td>\n",
       "10100\n",
       "</td>\n",
       "<td>\n",
       "&quot;C&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "12\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "110340653\n",
       "</td>\n",
       "<td>\n",
       "110340659\n",
       "</td>\n",
       "<td>\n",
       "&quot;CGGCTTT&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "13\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr11&quot;\n",
       "</td>\n",
       "<td>\n",
       "108227625\n",
       "</td>\n",
       "<td>\n",
       "108227625\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "362\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "124022628\n",
       "</td>\n",
       "<td>\n",
       "124022628\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "363\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "134377618\n",
       "</td>\n",
       "<td>\n",
       "134377618\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "364\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "153907795\n",
       "</td>\n",
       "<td>\n",
       "153907795\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;G&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "365\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "40064468\n",
       "</td>\n",
       "<td>\n",
       "40064468\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "366\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "44873624\n",
       "</td>\n",
       "<td>\n",
       "44873624\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "367\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "48791110\n",
       "</td>\n",
       "<td>\n",
       "48791110\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "368\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "53224889\n",
       "</td>\n",
       "<td>\n",
       "53224889\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "369\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "64193286\n",
       "</td>\n",
       "<td>\n",
       "64193286\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "370\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "67545147\n",
       "</td>\n",
       "<td>\n",
       "67545147\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "371\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "71120077\n",
       "</td>\n",
       "<td>\n",
       "71120077\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "372\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "77786001\n",
       "</td>\n",
       "<td>\n",
       "77786001\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "373\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrY&quot;\n",
       "</td>\n",
       "<td>\n",
       "13479665\n",
       "</td>\n",
       "<td>\n",
       "13479665\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "shape: (372, 6)\n",
       "┌─────┬───────┬──────────┬──────────┬──────────┬──────────┐\n",
       "│ uid ┆ chrom ┆ pos      ┆ pos_end  ┆ ref_base ┆ alt_base │\n",
       "│ --- ┆ ---   ┆ ---      ┆ ---      ┆ ---      ┆ ---      │\n",
       "│ i64 ┆ str   ┆ i64      ┆ i64      ┆ str      ┆ str      │\n",
       "╞═════╪═══════╪══════════╪══════════╪══════════╪══════════╡\n",
       "│ 2   ┆ chr10 ┆ 2987654  ┆ 2987654  ┆ T        ┆ A        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 3   ┆ chr10 ┆ 43077259 ┆ 43077259 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 4   ┆ chr10 ┆ 8055656  ┆ 8055656  ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 5   ┆ chr10 ┆ 87864470 ┆ 87864470 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ ... ┆ ...   ┆ ...      ┆ ...      ┆ ...      ┆ ...      │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 370 ┆ chrX  ┆ 67545147 ┆ 67545147 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 371 ┆ chrX  ┆ 71120077 ┆ 71120077 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 372 ┆ chrX  ┆ 77786001 ┆ 77786001 ┆ T        ┆ A        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 373 ┆ chrY  ┆ 13479665 ┆ 13479665 ┆ T        ┆ A        │\n",
       "└─────┴───────┴──────────┴──────────┴──────────┴──────────┘"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.run_df(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "127b1812-4258-4318-a164-453eef4dc592",
   "metadata": {},
   "outputs": [],
   "source": [
    "m.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dcf922e9-8325-4fd7-b98e-6d5d0623d643",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=ov.read_crv(\"exampleinput.crv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e4e65e63-d23c-415a-b767-af1ba3a1bcdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "f=open(\"exampleinput.crv\")\n",
    "c = 0\n",
    "for line in f:\n",
    "    if not line.startswith(\"#\"):\n",
    "        break\n",
    "    c += 1\n",
    "input_df=pl.read_csv(f, skip_rows=c, new_columns=[\"uid\", \"chrom\", \"pos\", \"pos_end\", \"ref_base\", \"alt_base\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7a757b0b-907d-49c4-83a5-df88c4d19586",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        white-space: pre;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-top: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        padding-bottom: 0;\n",
       "    }\n",
       "\n",
       "    .dataframe td {\n",
       "        line-height: 95%;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\" >\n",
       "<small>shape: (372, 6)</small>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>\n",
       "uid\n",
       "</th>\n",
       "<th>\n",
       "chrom\n",
       "</th>\n",
       "<th>\n",
       "pos\n",
       "</th>\n",
       "<th>\n",
       "pos_end\n",
       "</th>\n",
       "<th>\n",
       "ref_base\n",
       "</th>\n",
       "<th>\n",
       "alt_base\n",
       "</th>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "i64\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "<td>\n",
       "str\n",
       "</td>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>\n",
       "2\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "2987654\n",
       "</td>\n",
       "<td>\n",
       "2987654\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "3\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "43077259\n",
       "</td>\n",
       "<td>\n",
       "43077259\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "4\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "8055656\n",
       "</td>\n",
       "<td>\n",
       "8055656\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "5\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864470\n",
       "</td>\n",
       "<td>\n",
       "87864470\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "6\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "7\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87864486\n",
       "</td>\n",
       "<td>\n",
       "87864487\n",
       "</td>\n",
       "<td>\n",
       "&quot;AA&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "8\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;CG&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "9\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr10&quot;\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "87894027\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;CT&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "10\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "100719861\n",
       "</td>\n",
       "<td>\n",
       "100719861\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "11\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "10100\n",
       "</td>\n",
       "<td>\n",
       "10100\n",
       "</td>\n",
       "<td>\n",
       "&quot;C&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "12\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr1&quot;\n",
       "</td>\n",
       "<td>\n",
       "110340653\n",
       "</td>\n",
       "<td>\n",
       "110340659\n",
       "</td>\n",
       "<td>\n",
       "&quot;CGGCTTT&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "13\n",
       "</td>\n",
       "<td>\n",
       "&quot;chr11&quot;\n",
       "</td>\n",
       "<td>\n",
       "108227625\n",
       "</td>\n",
       "<td>\n",
       "108227625\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "<td>\n",
       "...\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "362\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "124022628\n",
       "</td>\n",
       "<td>\n",
       "124022628\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "363\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "134377618\n",
       "</td>\n",
       "<td>\n",
       "134377618\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "364\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "153907795\n",
       "</td>\n",
       "<td>\n",
       "153907795\n",
       "</td>\n",
       "<td>\n",
       "&quot;-&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;G&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "365\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "40064468\n",
       "</td>\n",
       "<td>\n",
       "40064468\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "366\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "44873624\n",
       "</td>\n",
       "<td>\n",
       "44873624\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "367\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "48791110\n",
       "</td>\n",
       "<td>\n",
       "48791110\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "368\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "53224889\n",
       "</td>\n",
       "<td>\n",
       "53224889\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "369\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "64193286\n",
       "</td>\n",
       "<td>\n",
       "64193286\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "370\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "67545147\n",
       "</td>\n",
       "<td>\n",
       "67545147\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "371\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "71120077\n",
       "</td>\n",
       "<td>\n",
       "71120077\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "372\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrX&quot;\n",
       "</td>\n",
       "<td>\n",
       "77786001\n",
       "</td>\n",
       "<td>\n",
       "77786001\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>\n",
       "373\n",
       "</td>\n",
       "<td>\n",
       "&quot;chrY&quot;\n",
       "</td>\n",
       "<td>\n",
       "13479665\n",
       "</td>\n",
       "<td>\n",
       "13479665\n",
       "</td>\n",
       "<td>\n",
       "&quot;T&quot;\n",
       "</td>\n",
       "<td>\n",
       "&quot;A&quot;\n",
       "</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "shape: (372, 6)\n",
       "┌─────┬───────┬──────────┬──────────┬──────────┬──────────┐\n",
       "│ uid ┆ chrom ┆ pos      ┆ pos_end  ┆ ref_base ┆ alt_base │\n",
       "│ --- ┆ ---   ┆ ---      ┆ ---      ┆ ---      ┆ ---      │\n",
       "│ i64 ┆ str   ┆ i64      ┆ i64      ┆ str      ┆ str      │\n",
       "╞═════╪═══════╪══════════╪══════════╪══════════╪══════════╡\n",
       "│ 2   ┆ chr10 ┆ 2987654  ┆ 2987654  ┆ T        ┆ A        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 3   ┆ chr10 ┆ 43077259 ┆ 43077259 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 4   ┆ chr10 ┆ 8055656  ┆ 8055656  ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 5   ┆ chr10 ┆ 87864470 ┆ 87864470 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ ... ┆ ...   ┆ ...      ┆ ...      ┆ ...      ┆ ...      │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 370 ┆ chrX  ┆ 67545147 ┆ 67545147 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 371 ┆ chrX  ┆ 71120077 ┆ 71120077 ┆ A        ┆ T        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 372 ┆ chrX  ┆ 77786001 ┆ 77786001 ┆ T        ┆ A        │\n",
       "├╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┤\n",
       "│ 373 ┆ chrY  ┆ 13479665 ┆ 13479665 ┆ T        ┆ A        │\n",
       "└─────┴───────┴──────────┴──────────┴──────────┴──────────┘"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "27b0c663-14e2-4a65-991e-b5ad17d73353",
   "metadata": {},
   "outputs": [],
   "source": [
    "mc = ov.MasterConverter(inputs=[\"exampleinput\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9af71154-92f5-4ba8-9d13-4fc775e1947f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'total_lnum': 373,\n",
       " 'write_lnum': 373,\n",
       " 'error_lnum': 0,\n",
       " 'input_format': ['cravat'],\n",
       " 'assemblies': ['hg38']}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6d30db40-527c-4bfc-a8ad-19137d027bd2",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "<class '__main__.Annotator'> is a built-in class",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [43]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m inspect\u001b[38;5;241m.\u001b[39mlinecache\u001b[38;5;241m.\u001b[39mgetlines(\u001b[43minspect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetfile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mAnnotator\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:665\u001b[0m, in \u001b[0;36mgetfile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    663\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(module, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__file__\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    664\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m module\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__file__\u001b[39m\n\u001b[0;32m--> 665\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{!r}\u001b[39;00m\u001b[38;5;124m is a built-in class\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\u001b[38;5;28mobject\u001b[39m))\n\u001b[1;32m    666\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ismethod(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    667\u001b[0m     \u001b[38;5;28mobject\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__func__\u001b[39m\n",
      "\u001b[0;31mTypeError\u001b[0m: <class '__main__.Annotator'> is a built-in class"
     ]
    }
   ],
   "source": [
    "inspect.linecache.getlines(inspect.getfile(Annotator))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4d38d8ec-046d-47c5-b959-3fcd06d61b25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class Annotator(ov.BaseAnnotator):\n",
      "    def annotate(self, input_data: dict):\n",
      "        out_data = input_data\n",
      "        out_data[\"col1\"] = \"Test\"\n",
      "        return out_data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"\".join(inspect.getsourcelines(Annotator)[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "76b5de03-eecf-4e2f-95e1-54c114544e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.magics.code import extract_symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6fecee13-7521-4fea-a4f8-204be84632ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_code = \"\".join(inspect.linecache.getlines(new_getfile(Annotator)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e04fea9d-6b50-42d6-8de5-df49db3a5bd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'import oakvar as ov\\nclass Annotator(ov.BaseAnnotator):\\n    def annotate(self, input_data: dict):\\n        out_data = input_data\\n        out_data[\"col1\"] = \"Test\"\\n        return out_data\\nm = Annotator(module_name=\"awesome\", level=\"variant\", output_columns=[{\"name\": \"col1\", \"type\": \"string\", \"title\": \"Col1\"}])\\n'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dae52793-fb51-4eff-98ad-4e1e418825ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@@@ module_conf={}. output_columns=[{'name': 'col1', 'type': 'string', 'title': 'Col1'}]\n",
      "@@@ conf={}\n",
      "@ 0 self.conf={}\n",
      "@ output_columns=[{'name': 'col1', 'type': 'string', 'title': 'Col1'}]. self.output_columns=None. self.conf={'level': 'variant'}. module_conf={}\n",
      "@@ self.output_columns=[{'name': 'col1', 'type': 'string', 'title': 'Col1'}]\n",
      "@ conf={'level': 'variant', 'output_columns': [{'name': 'col1', 'type': 'string', 'title': 'Col1'}]}\n",
      "@ => conf={'level': 'variant', 'output_columns': [{'name': 'uid', 'title': 'UID', 'type': 'int', 'width': 60, 'hidden': True, 'filterable': False}, {'name': 'col1', 'type': 'string', 'title': 'Col1'}]}\n"
     ]
    }
   ],
   "source": [
    "m = Annotator(module_name=\"awesome\", level=\"variant\", output_columns=[{\"name\": \"col1\", \"type\": \"string\", \"title\": \"Col1\"}])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ecc2d5a6-7cd8-4bec-bbb9-dcfb83d24a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"text.py\", \"w\") as wf:\n",
    "    wf.write(str(Annotator))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bd2139bc-2d52-4d14-9f5e-f0450df5a369",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'type'>\n"
     ]
    }
   ],
   "source": [
    "print(type(Annotator))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f1d73e49-e8ac-4a06-9742-fe617fdfd3d1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['class BaseAnnotator(object):\\n',\n",
       "  '\\n',\n",
       "  '    from ..util.util import get_crv_def\\n',\n",
       "  '    from ..util.util import get_crx_def\\n',\n",
       "  '    from ..util.util import get_crg_def\\n',\n",
       "  '    from ..consts import INPUT_LEVEL_KEY\\n',\n",
       "  '    from ..consts import VARIANT_LEVEL_KEY\\n',\n",
       "  '    from ..consts import GENE_LEVEL_KEY\\n',\n",
       "  '\\n',\n",
       "  '    valid_levels = [\"variant\", \"gene\"]\\n',\n",
       "  '    valid_input_formats = [INPUT_LEVEL_KEY, VARIANT_LEVEL_KEY, GENE_LEVEL_KEY]\\n',\n",
       "  '    id_col_defs = {\"variant\": get_crv_def()[0], \"gene\": get_crg_def()[0]}\\n',\n",
       "  '    default_input_columns = {\\n',\n",
       "  '        INPUT_LEVEL_KEY: [x[\"name\"] for x in get_crv_def()],\\n',\n",
       "  '        VARIANT_LEVEL_KEY: [x[\"name\"] for x in get_crx_def()],\\n',\n",
       "  '        GENE_LEVEL_KEY: [x[\"name\"] for x in get_crg_def()],\\n',\n",
       "  '    }\\n',\n",
       "  '    required_conf_keys = [\"level\", \"output_columns\"]\\n',\n",
       "  '\\n',\n",
       "  '    def __init__(self, input_file: Optional[str]=None, secondary_inputs=None, run_name: Optional[str]=None, output_dir: Optional[str]=None, plainoutput: bool=False, logtofile: bool=False, module_options: Dict={}, serveradmindb=None, module_name: Optional[str]=None, title: Optional[str]=None, level: Optional[str]=None, output_columns: List[Dict]=[], module_conf: dict={}):\\n',\n",
       "  '        import os\\n',\n",
       "  '        import sys\\n',\n",
       "  '        from pathlib import Path\\n',\n",
       "  '        from ..consts import cannonical_chroms\\n',\n",
       "  '        from ..module.local import get_module_conf\\n',\n",
       "  '        from ..module.data_cache import ModuleDataCache\\n',\n",
       "  '        from ..exceptions import ModuleLoadingError\\n',\n",
       "  '        from ..exceptions import LoggerError\\n',\n",
       "  '\\n',\n",
       "  '        self.module_options = module_options\\n',\n",
       "  '        if input_file:\\n',\n",
       "  '            self.primary_input_path = Path(input_file).absolute()\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.primary_input_path = None\\n',\n",
       "  '        self.secondary_inputs = secondary_inputs\\n',\n",
       "  '        self.run_name = run_name\\n',\n",
       "  '        self.output_dir = output_dir\\n',\n",
       "  '        self.plain_output = plainoutput\\n',\n",
       "  '        self.logtofile = logtofile\\n',\n",
       "  '        if self.__module__ == \"__main__\":\\n',\n",
       "  '            fp = None\\n',\n",
       "  '            self.main_fpath = None\\n',\n",
       "  '        else:\\n',\n",
       "  '            fp = sys.modules[self.__module__].__file__\\n',\n",
       "  '            if not fp:\\n',\n",
       "  '                raise ModuleLoadingError(module_name=self.__module__)\\n',\n",
       "  '            self.main_fpath = Path(fp).resolve()\\n',\n",
       "  '        self.secondary_paths = {}\\n',\n",
       "  '        self.output_basename = None\\n',\n",
       "  '        self.job_conf_path = None\\n',\n",
       "  '        self.logger = None\\n',\n",
       "  '        self.error_logger = None\\n',\n",
       "  '        self.dbconn = None\\n',\n",
       "  '        self.cursor = None\\n',\n",
       "  '        self.cmd_arg_parser = None\\n',\n",
       "  '        self.json_colnames = None\\n',\n",
       "  '        self.primary_input_reader = None\\n',\n",
       "  '        self.output_path = None\\n',\n",
       "  '        self.last_status_update_time = None\\n',\n",
       "  '        self.output_columns = None\\n',\n",
       "  '        self.secondary_readers = {}\\n',\n",
       "  '        self.output_writer = None\\n',\n",
       "  '        self.log_path = None\\n',\n",
       "  '        self.unique_excs = []\\n',\n",
       "  '        self.log_handler = None\\n',\n",
       "  '        self.parse_cmd_args()\\n',\n",
       "  '        self.serveradmindb = serveradmindb\\n',\n",
       "  '        self.supported_chroms = set(cannonical_chroms)\\n',\n",
       "  '        self.module_type = \"annotator\"\\n',\n",
       "  '        if not self.main_fpath:\\n',\n",
       "  '            if module_name:\\n',\n",
       "  '                self.module_name = module_name\\n',\n",
       "  '                self.module_dir = Path(os.getcwd()).absolute()\\n',\n",
       "  '                self.annotator_dir = Path(os.getcwd()).absolute()\\n',\n",
       "  '            else:\\n',\n",
       "  '                raise ModuleLoadingError(msg=\"module_name argument should be given.\")\\n',\n",
       "  '            self.conf = module_conf.copy()\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.module_name = self.main_fpath.stem\\n',\n",
       "  '            self.module_dir = self.main_fpath.parent\\n',\n",
       "  '            self.annotator_dir = self.main_fpath.parent\\n',\n",
       "  '            self.conf = get_module_conf(\\n',\n",
       "  '                self.module_name, module_type=self.module_type, module_dir=self.module_dir\\n',\n",
       "  '            )\\n',\n",
       "  '        self.level = level\\n',\n",
       "  '        if not self.level and self.conf and \"level\" in self.conf:\\n',\n",
       "  '            self.level = self.conf.get(\"level\")\\n',\n",
       "  '        if not self.level:\\n',\n",
       "  '            raise ModuleLoadingError(msg=\"level or module_conf with level should be given.\")\\n',\n",
       "  '        if self.conf is not None and \"level\" not in self.conf:\\n',\n",
       "  '            self.conf[\"level\"] = self.level\\n',\n",
       "  '        self.output_columns = output_columns.copy()\\n',\n",
       "  '        if self.output_columns is not None and self.conf is not None and \"output_columns\" not in self.conf:\\n',\n",
       "  '            self.conf[\"output_columns\"] = self.output_columns\\n',\n",
       "  '        elif not self.output_columns and self.conf and \"output_columns\" in self.conf:\\n',\n",
       "  '            self.output_columns = self.conf[\"output_columns\"]\\n',\n",
       "  '        self.title = title\\n',\n",
       "  '        if self.title and self.conf is not None:\\n',\n",
       "  '            self.conf[\"title\"] = self.title\\n',\n",
       "  '        elif self.conf and \"title\" in self.conf:\\n',\n",
       "  '            self.title = self.conf[\"title\"]\\n',\n",
       "  '        self.annotator_name = self.module_name\\n',\n",
       "  '        self.data_dir = self.module_dir / \"data\"\\n',\n",
       "  '        self._setup_logger()\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            raise ModuleLoadingError(module_name=self.module_name)\\n',\n",
       "  '        self.set_ref_colname()\\n',\n",
       "  '        self._verify_conf()\\n',\n",
       "  '        if self.logger is None:\\n',\n",
       "  '            raise LoggerError(module_name=self.module_name)\\n',\n",
       "  '        if \"logging_level\" in self.conf:\\n',\n",
       "  '            self.logger.setLevel(self.conf[\"logging_level\"].upper())\\n',\n",
       "  '        if \"title\" in self.conf:\\n',\n",
       "  '            self.annotator_display_name = self.conf[\"title\"]\\n',\n",
       "  '        elif self.module_name:\\n',\n",
       "  '            self.annotator_display_name = self.module_name.capitalize()\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.annotator_display_name = os.path.basename(self.module_dir).upper()\\n',\n",
       "  '        if \"version\" in self.conf:\\n',\n",
       "  '            self.annotator_version = self.conf[\"version\"]\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.annotator_version = \"\"\\n',\n",
       "  '        self.cache = ModuleDataCache(self.module_name, module_type=self.module_type)\\n',\n",
       "  '\\n',\n",
       "  '    def set_ref_colname(self):\\n',\n",
       "  '        ref_colnames = {\"variant\": \"uid\", \"gene\": \"hugo\", \"sample\": \"uid\", \"mapping\": \"uid\"}\\n',\n",
       "  '        if self.level:\\n',\n",
       "  '            self._id_col_name = ref_colnames.get(self.level)\\n',\n",
       "  '        else:\\n',\n",
       "  '            self._id_col_name = None\\n',\n",
       "  '\\n',\n",
       "  '    def summarize_by_gene(self, __hugo__, __input_data__):\\n',\n",
       "  '        pass\\n',\n",
       "  '\\n',\n",
       "  '    def _log_exception(self, e, halt=True):\\n',\n",
       "  '        import traceback\\n',\n",
       "  '\\n',\n",
       "  '        if self.logger:\\n',\n",
       "  '            self.logger.exception(e)\\n',\n",
       "  '        else:\\n',\n",
       "  '            tb = traceback.format_exc()\\n',\n",
       "  '            print(tb)\\n',\n",
       "  '        if halt:\\n',\n",
       "  '            return False\\n',\n",
       "  '        else:\\n',\n",
       "  '            return True\\n',\n",
       "  '\\n',\n",
       "  '    def _verify_conf(self):\\n',\n",
       "  '        from ..consts import VARIANT_LEVEL_KEY\\n',\n",
       "  '        from ..consts import GENE_LEVEL_KEY\\n',\n",
       "  '\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            from ..exceptions import SetupError\\n',\n",
       "  '\\n',\n",
       "  '            raise SetupError(self.module_name)\\n',\n",
       "  '        from ..exceptions import ConfigurationError\\n',\n",
       "  '\\n',\n",
       "  '        for k in self.required_conf_keys:\\n',\n",
       "  '            if k not in self.conf:\\n',\n",
       "  '                err_msg = \\'Required key \"%s\" not found in configuration\\' % k\\n',\n",
       "  '                raise ConfigurationError(err_msg)\\n',\n",
       "  '        if self.conf[\"level\"] in self.valid_levels and self.level:\\n',\n",
       "  '            self.conf[\"output_columns\"] = [\\n',\n",
       "  '                self.id_col_defs[self.level]\\n',\n",
       "  '            ] + self.conf[\"output_columns\"]\\n',\n",
       "  '        else:\\n',\n",
       "  '            err_msg = \"%s is not a valid level. Valid levels are %s\" % (\\n',\n",
       "  '                self.conf[\"level\"],\\n',\n",
       "  '                \", \".join(self.valid_levels),\\n',\n",
       "  '            )\\n',\n",
       "  '            raise ConfigurationError(err_msg)\\n',\n",
       "  '        if \"input_format\" in self.conf:\\n',\n",
       "  '            if self.conf[\"input_format\"] not in self.valid_input_formats:\\n',\n",
       "  '                err_msg = \"Invalid input_format %s, select from %s\" % (\\n',\n",
       "  '                    self.conf[\"input_format\"],\\n',\n",
       "  '                    \", \".join(self.valid_input_formats),\\n',\n",
       "  '                )\\n',\n",
       "  '                raise ConfigurationError(err_msg)\\n',\n",
       "  '        else:\\n',\n",
       "  '            if self.conf[\"level\"] == \"variant\":\\n',\n",
       "  '                self.conf[\"input_format\"] = VARIANT_LEVEL_KEY\\n',\n",
       "  '            elif self.conf[\"level\"] == \"gene\":\\n',\n",
       "  '                self.conf[\"input_format\"] = GENE_LEVEL_KEY\\n',\n",
       "  '        if \"input_columns\" in self.conf:\\n',\n",
       "  '            id_col_name = self.id_col_defs[self.conf[\"level\"]][\"name\"]\\n',\n",
       "  '            if id_col_name not in self.conf[\"input_columns\"]:\\n',\n",
       "  '                self.conf[\"input_columns\"].append(id_col_name)\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.conf[\"input_columns\"] = self.default_input_columns[\\n',\n",
       "  '                self.conf[\"input_format\"]\\n',\n",
       "  '            ]\\n',\n",
       "  '\\n',\n",
       "  '    def parse_cmd_args(self):\\n',\n",
       "  '        import re\\n',\n",
       "  '        from pathlib import Path\\n',\n",
       "  '\\n',\n",
       "  '        if self.secondary_inputs:\\n',\n",
       "  '            for secondary_def in self.secondary_inputs:\\n',\n",
       "  '                sec_name, sec_path = re.split(r\"(?<!\\\\\\\\)=\", secondary_def)\\n',\n",
       "  '                self.secondary_paths[sec_name] = str(Path(sec_path).absolute())\\n',\n",
       "  '        if not self.output_dir and self.primary_input_path:\\n',\n",
       "  '            self.output_dir = str(self.primary_input_path.parent)\\n',\n",
       "  '        if self.run_name is not None:\\n',\n",
       "  '            self.output_basename = self.run_name\\n',\n",
       "  '        elif self.primary_input_path:\\n',\n",
       "  '            self.output_basename = self.primary_input_path.name\\n',\n",
       "  '            if Path(self.output_basename).suffix in [\".crv\", \".crg\", \".crx\"]:\\n',\n",
       "  '                self.output_basename = self.output_basename[:-4]\\n',\n",
       "  '\\n',\n",
       "  '    def handle_jsondata(self, output_dict):\\n',\n",
       "  '        import json\\n',\n",
       "  '\\n',\n",
       "  '        if self.json_colnames is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        for colname in self.json_colnames:\\n',\n",
       "  '            json_data = output_dict.get(colname, None)\\n',\n",
       "  '            if json_data is not None:\\n',\n",
       "  '                json_data = json.dumps(json_data)\\n',\n",
       "  '            output_dict[colname] = json_data\\n',\n",
       "  '        return output_dict\\n',\n",
       "  '\\n',\n",
       "  '    def log_progress(self, lnum):\\n',\n",
       "  '        from time import time\\n',\n",
       "  '        from ..util.run import update_status\\n',\n",
       "  '        from ..consts import JOB_STATUS_UPDATE_INTERVAL\\n',\n",
       "  '\\n',\n",
       "  '        if not self.last_status_update_time or not self.conf:\\n',\n",
       "  '            return\\n',\n",
       "  '        cur_time = time()\\n',\n",
       "  '        if (\\n',\n",
       "  '            lnum % 10000 == 0\\n',\n",
       "  '            or cur_time - self.last_status_update_time > JOB_STATUS_UPDATE_INTERVAL\\n',\n",
       "  '        ):\\n',\n",
       "  '            status = f\"Running {self.conf[\\'title\\']} ({self.module_name}): line {lnum}\"\\n',\n",
       "  '            update_status(status, logger=self.logger, serveradmindb=self.serveradmindb)\\n',\n",
       "  '            self.last_status_update_time = cur_time\\n',\n",
       "  '\\n',\n",
       "  '    def is_star_allele(self, input_data):\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        return self.conf[\"level\"] == \"variant\" and input_data.get(\"alt_base\", \"\") == \"*\"\\n',\n",
       "  '\\n',\n",
       "  '    def should_skip_chrom(self, input_data):\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        return (\\n',\n",
       "  '            self.conf[\"level\"] == \"variant\"\\n',\n",
       "  '            and not input_data.get(\"chrom\") in self.supported_chroms\\n',\n",
       "  '        )\\n',\n",
       "  '\\n',\n",
       "  '    def fill_empty_output(self, output_dict):\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        for output_col in self.conf[\"output_columns\"]:\\n',\n",
       "  '            col_name = output_col[\"name\"]\\n',\n",
       "  '            if col_name not in output_dict:\\n',\n",
       "  '                output_dict[col_name] = None\\n',\n",
       "  '        return output_dict\\n',\n",
       "  '\\n',\n",
       "  '    def make_json_colnames(self):\\n',\n",
       "  '        if self.output_columns is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        self.json_colnames = []\\n',\n",
       "  '        for col in self.output_columns:\\n',\n",
       "  '            if \"table\" in col and col[\"table\"] == True:\\n',\n",
       "  '                self.json_colnames.append(col[\"name\"])\\n',\n",
       "  '\\n',\n",
       "  '    # Runs the annotator.\\n',\n",
       "  '    def run(self):\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        if self.logger is None:\\n',\n",
       "  '            return\\n',\n",
       "  '        from time import time, asctime, localtime\\n',\n",
       "  '        from ..util.run import update_status\\n',\n",
       "  '        from ..exceptions import ModuleLoadingError\\n',\n",
       "  '\\n',\n",
       "  '        if not self.module_name:\\n',\n",
       "  '            raise ModuleLoadingError(msg=\"module_name should be given at initializing Annotator to run.\")\\n',\n",
       "  '        if self.conf:\\n',\n",
       "  '            if \"title\" not in self.conf:\\n',\n",
       "  '                raise ModuleLoadingError(msg=\"title should be given at initializing Annotator or in the module yml file to run.\")\\n',\n",
       "  '            if \"level\" not in self.conf:\\n',\n",
       "  '                raise ModuleLoadingError(msg=\"level should be given at initializing Annotator or in the module yml file to run.\")\\n',\n",
       "  '            if \"output_columns\" not in self.conf:\\n',\n",
       "  '                raise ModuleLoadingError(msg=\"output_columns should be given at initializing Annotator or in the module yml file to run.\")\\n',\n",
       "  '            if not self.primary_input_path:\\n',\n",
       "  '                raise ModuleLoadingError(msg=\"input_file should be given at initializing Annotator to run.\")\\n',\n",
       "  '        else:\\n',\n",
       "  '            raise ModuleLoadingError(msg=\"module conf should exist to run.\")\\n',\n",
       "  '        status = f\"started {self.conf[\\'title\\']} ({self.module_name})\"\\n',\n",
       "  '        update_status(status, logger=self.logger, serveradmindb=self.serveradmindb)\\n',\n",
       "  '        try:\\n',\n",
       "  '            start_time = time()\\n',\n",
       "  '            update_status(\\n',\n",
       "  '                \"started: %s\" % asctime(localtime(start_time)),\\n',\n",
       "  '                logger=self.logger,\\n',\n",
       "  '                serveradmindb=self.serveradmindb,\\n',\n",
       "  '            )\\n',\n",
       "  '            self.base_setup()\\n',\n",
       "  '            self.last_status_update_time = time()\\n',\n",
       "  '            self.output_columns = self.conf[\"output_columns\"]\\n',\n",
       "  '            self.make_json_colnames()\\n',\n",
       "  '            self.process_file()\\n',\n",
       "  '            self.postprocess()\\n',\n",
       "  '            self.base_cleanup()\\n',\n",
       "  '            status = f\"started {self.conf[\\'title\\']} ({self.module_name})\"\\n',\n",
       "  '            update_status(status, logger=self.logger, serveradmindb=self.serveradmindb)\\n',\n",
       "  '            end_time = time()\\n',\n",
       "  '            update_status(\\n',\n",
       "  '                f\"{self.module_name}: finished at {asctime(localtime(end_time))}\",\\n',\n",
       "  '                logger=self.logger,\\n',\n",
       "  '                serveradmindb=self.serveradmindb,\\n',\n",
       "  '            )\\n',\n",
       "  '            run_time = end_time - start_time\\n',\n",
       "  '            update_status(\\n',\n",
       "  '                f\"{self.module_name}: runtime {run_time:0.3f}s\",\\n',\n",
       "  '                logger=self.logger,\\n',\n",
       "  '                serveradmindb=self.serveradmindb,\\n',\n",
       "  '            )\\n',\n",
       "  '        except Exception as e:\\n',\n",
       "  '            self._log_exception(e)\\n',\n",
       "  '        if hasattr(self, \"log_handler\") and self.log_handler:\\n',\n",
       "  '            self.log_handler.close()\\n',\n",
       "  '\\n',\n",
       "  '    def process_file(self):\\n',\n",
       "  '        assert self._id_col_name, \"_id_col_name should not be None.\"\\n',\n",
       "  '        for lnum, line, input_data, secondary_data in self._get_input():\\n',\n",
       "  '            try:\\n',\n",
       "  '                self.log_progress(lnum)\\n',\n",
       "  '                # * allele and undefined non-canonical chroms are skipped.\\n',\n",
       "  '                if self.is_star_allele(input_data) or self.should_skip_chrom(\\n',\n",
       "  '                    input_data\\n',\n",
       "  '                ):\\n',\n",
       "  '                    continue\\n',\n",
       "  '                output_dict = None\\n',\n",
       "  '                if secondary_data == {}:\\n',\n",
       "  '                    output_dict = self.annotate(input_data)\\n',\n",
       "  '                else:\\n',\n",
       "  '                    output_dict = self.annotate(\\n',\n",
       "  '                        input_data, secondary_data=secondary_data\\n',\n",
       "  '                    )\\n',\n",
       "  '                # This enables summarizing without writing for now.\\n',\n",
       "  '                if output_dict is None:\\n',\n",
       "  '                    continue\\n',\n",
       "  '                # Handles empty table-format column data.\\n',\n",
       "  '                output_dict = self.handle_jsondata(output_dict)\\n',\n",
       "  '                # Preserves the first column\\n',\n",
       "  '                if output_dict:\\n',\n",
       "  '                    output_dict[self._id_col_name] = input_data[self._id_col_name]\\n',\n",
       "  '                # Fill absent columns with empty strings\\n',\n",
       "  '                output_dict = self.fill_empty_output(output_dict)\\n',\n",
       "  '                # Writes output.\\n',\n",
       "  '                if self.output_writer:\\n',\n",
       "  '                    self.output_writer.write_data(output_dict)\\n',\n",
       "  '            except Exception as e:\\n',\n",
       "  '                self._log_runtime_exception(\\n',\n",
       "  '                    lnum,\\n',\n",
       "  '                    line,\\n',\n",
       "  '                    input_data,\\n',\n",
       "  '                    e,\\n',\n",
       "  '                    fn=self.primary_input_reader.path\\n',\n",
       "  '                    if self.primary_input_reader\\n',\n",
       "  '                    else \"?\",\\n',\n",
       "  '                )\\n',\n",
       "  '\\n',\n",
       "  '    def postprocess(self):\\n',\n",
       "  '        pass\\n',\n",
       "  '\\n',\n",
       "  '    async def get_gene_summary_data(self, cf):\\n',\n",
       "  '        hugos = await cf.exec_db(cf.get_filtered_hugo_list)\\n',\n",
       "  '        output_columns = await cf.exec_db(\\n',\n",
       "  '            cf.get_stored_output_columns, self.module_name\\n',\n",
       "  '        )\\n',\n",
       "  '        cols = [\\n',\n",
       "  '            self.module_name + \"__\" + coldef[\"name\"]\\n',\n",
       "  '            for coldef in output_columns\\n',\n",
       "  '            if coldef[\"name\"] != \"uid\"\\n',\n",
       "  '        ]\\n',\n",
       "  '        data = {}\\n',\n",
       "  '        rows = await cf.exec_db(cf.get_variant_data_for_cols, cols)\\n',\n",
       "  '        rows_by_hugo = {}\\n',\n",
       "  '        for row in rows:\\n',\n",
       "  '            hugo = row[-1]\\n',\n",
       "  '            if hugo not in rows_by_hugo:\\n',\n",
       "  '                rows_by_hugo[hugo] = []\\n',\n",
       "  '            rows_by_hugo[hugo].append(row)\\n',\n",
       "  '        for hugo in hugos:\\n',\n",
       "  '            rows = rows_by_hugo[hugo]\\n',\n",
       "  '            input_data = {}\\n',\n",
       "  '            for i in range(len(cols)):\\n',\n",
       "  '                input_data[cols[i].split(\"__\")[1]] = [row[i] for row in rows]\\n',\n",
       "  '            out = self.summarize_by_gene(hugo, input_data)\\n',\n",
       "  '            data[hugo] = out\\n',\n",
       "  '        return data\\n',\n",
       "  '\\n',\n",
       "  '    def _log_runtime_exception(self, lnum, __line__, __input_data__, e, fn=None):\\n',\n",
       "  '        import traceback\\n',\n",
       "  '\\n',\n",
       "  '        err_str = traceback.format_exc().rstrip()\\n',\n",
       "  '        lines = err_str.split(\"\\\\n\")\\n',\n",
       "  '        last_line = lines[-1]\\n',\n",
       "  '        err_str_log = \"\\\\n\".join(lines[:-1]) + \"\\\\n\" + \":\".join(last_line.split(\":\")[:2])\\n',\n",
       "  '        if err_str_log not in self.unique_excs:\\n',\n",
       "  '            self.unique_excs.append(err_str_log)\\n',\n",
       "  '            if self.logger:\\n',\n",
       "  '                self.logger.error(err_str_log)\\n',\n",
       "  '            else:\\n',\n",
       "  '                print(err_str_log)\\n',\n",
       "  '        err_logger_s = f\"{fn}:{lnum}\\\\t{str(e)}\"\\n',\n",
       "  '        if self.error_logger:\\n',\n",
       "  '            self.error_logger.error(err_logger_s)\\n',\n",
       "  '        else:\\n',\n",
       "  '            print(err_logger_s)\\n',\n",
       "  '\\n',\n",
       "  '    def base_setup(self):\\n',\n",
       "  '        self._setup_primary_input()\\n',\n",
       "  '        self._setup_secondary_inputs()\\n',\n",
       "  '        self._setup_outputs()\\n',\n",
       "  '        self.connect_db()\\n',\n",
       "  '        self.setup()\\n',\n",
       "  '        if not hasattr(self, \"supported_chroms\"):\\n',\n",
       "  '            self.supported_chroms = set(\\n',\n",
       "  '                [\"chr\" + str(n) for n in range(1, 23)] + [\"chrX\", \"chrY\"]\\n',\n",
       "  '            )\\n',\n",
       "  '\\n',\n",
       "  '    def _setup_primary_input(self):\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            from ..exceptions import SetupError\\n',\n",
       "  '\\n',\n",
       "  '            raise SetupError(module_name=self.module_name)\\n',\n",
       "  '        from ..exceptions import ConfigurationError\\n',\n",
       "  '        from ..util.inout import FileReader\\n',\n",
       "  '\\n',\n",
       "  '        self.primary_input_reader = FileReader(str(self.primary_input_path))\\n',\n",
       "  '        requested_input_columns = self.conf[\"input_columns\"]\\n',\n",
       "  '        defined_columns = self.primary_input_reader.get_column_names()\\n',\n",
       "  '        missing_columns = set(requested_input_columns) - set(defined_columns)\\n',\n",
       "  '        if missing_columns:\\n',\n",
       "  '            if len(defined_columns) > 0:\\n',\n",
       "  '                err_msg = \"Columns not defined in input: %s\" % \", \".join(\\n',\n",
       "  '                    missing_columns\\n',\n",
       "  '                )\\n',\n",
       "  '                raise ConfigurationError(err_msg)\\n',\n",
       "  '            else:\\n',\n",
       "  '                default_columns = self.default_input_columns[self.conf[\"input_format\"]]\\n',\n",
       "  '                for col_name in requested_input_columns:\\n',\n",
       "  '                    try:\\n',\n",
       "  '                        col_index = default_columns.index(col_name)\\n',\n",
       "  '                    except ValueError:\\n',\n",
       "  '                        err_msg = \"Column %s not defined for %s format input\" % (\\n',\n",
       "  '                            col_name,\\n',\n",
       "  '                            self.conf[\"input_format\"],\\n',\n",
       "  '                        )\\n',\n",
       "  '                        raise ConfigurationError(err_msg)\\n',\n",
       "  '                    if col_name == \"pos\":\\n',\n",
       "  '                        data_type = \"int\"\\n',\n",
       "  '                    else:\\n',\n",
       "  '                        data_type = \"string\"\\n',\n",
       "  '                    self.primary_input_reader.override_column(\\n',\n",
       "  '                        col_index, col_name, data_type=data_type\\n',\n",
       "  '                    )\\n',\n",
       "  '\\n',\n",
       "  '    def _setup_secondary_inputs(self):\\n',\n",
       "  '        from ..exceptions import SetupError\\n',\n",
       "  '\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            raise SetupError(module_name=self.module_name)\\n',\n",
       "  '        self.secondary_readers = {}\\n',\n",
       "  '        try:\\n',\n",
       "  '            num_expected = len(self.conf[\"secondary_inputs\"])\\n',\n",
       "  '        except KeyError:\\n',\n",
       "  '            num_expected = 0\\n',\n",
       "  '        num_provided = len(self.secondary_paths)\\n',\n",
       "  '        if num_expected > num_provided:\\n',\n",
       "  '            raise Exception(\\n',\n",
       "  '                f\"Too few secondary inputs. {num_expected} expected, {num_provided} provided\"\\n',\n",
       "  '            )\\n',\n",
       "  '        elif num_expected < num_provided:\\n',\n",
       "  '            raise Exception(\\n',\n",
       "  '                \"Too many secondary inputs. %d expected, %d provided\"\\n',\n",
       "  '                % (num_expected, num_provided)\\n',\n",
       "  '            )\\n',\n",
       "  '        for sec_name, sec_input_path in self.secondary_paths.items():\\n',\n",
       "  '            key_col = (\\n',\n",
       "  '                self.conf[\"secondary_inputs\"][sec_name]\\n',\n",
       "  '                .get(\"match_columns\", {})\\n',\n",
       "  '                .get(\"secondary\", \"uid\")\\n',\n",
       "  '            )\\n',\n",
       "  '            use_columns = self.conf[\"secondary_inputs\"][sec_name].get(\"use_columns\", [])\\n',\n",
       "  '            fetcher = SecondaryInputFetcher(\\n',\n",
       "  '                sec_input_path, key_col, fetch_cols=use_columns\\n',\n",
       "  '            )\\n',\n",
       "  '            self.secondary_readers[sec_name] = fetcher\\n',\n",
       "  '\\n',\n",
       "  '    def _setup_outputs(self):\\n',\n",
       "  '        from os import makedirs\\n',\n",
       "  '        from pathlib import Path\\n',\n",
       "  '        from ..util.inout import FileWriter\\n',\n",
       "  '        from ..exceptions import SetupError\\n',\n",
       "  '        from ..consts import VARIANT_LEVEL_OUTPUT_SUFFIX\\n',\n",
       "  '        from ..consts import GENE_LEVEL_OUTPUT_SUFFIX\\n',\n",
       "  '\\n',\n",
       "  '        if self.conf is None or self.output_dir is None or self.output_basename is None:\\n',\n",
       "  '            raise SetupError(module_name=self.module_name)\\n',\n",
       "  '        level = self.conf[\"level\"]\\n',\n",
       "  '        if level == \"variant\":\\n',\n",
       "  '            output_suffix = VARIANT_LEVEL_OUTPUT_SUFFIX\\n',\n",
       "  '        elif level == \"gene\":\\n',\n",
       "  '            output_suffix = GENE_LEVEL_OUTPUT_SUFFIX\\n',\n",
       "  '        elif level == \"summary\":\\n',\n",
       "  '            output_suffix = \".sum\"\\n',\n",
       "  '        else:\\n',\n",
       "  '            output_suffix = \".out\"\\n',\n",
       "  '        if not (Path(self.output_dir).exists):\\n',\n",
       "  '            makedirs(self.output_dir)\\n',\n",
       "  '        self.output_path = Path(self.output_dir) / f\"{self.output_basename}.{self.module_name}{output_suffix}\"\\n',\n",
       "  '        if self.plain_output:\\n',\n",
       "  '            self.output_writer = FileWriter(\\n',\n",
       "  '                self.output_path,\\n',\n",
       "  '                include_definition=False,\\n',\n",
       "  '                include_titles=True,\\n',\n",
       "  '                titles_prefix=\"\",\\n',\n",
       "  '            )\\n',\n",
       "  '        else:\\n',\n",
       "  '            self.output_writer = FileWriter(self.output_path)\\n',\n",
       "  '            self.output_writer.write_meta_line(\"name\", self.module_name)\\n',\n",
       "  '            self.output_writer.write_meta_line(\\n',\n",
       "  '                \"displayname\", self.annotator_display_name\\n',\n",
       "  '            )\\n',\n",
       "  '            self.output_writer.write_meta_line(\"version\", self.annotator_version)\\n',\n",
       "  '        skip_aggregation = []\\n',\n",
       "  '        for col_def in self.conf[\"output_columns\"]:\\n',\n",
       "  '            self.output_writer.add_column(col_def)\\n',\n",
       "  '            if not (col_def.get(\"aggregate\", True)):\\n',\n",
       "  '                skip_aggregation.append(col_def[\"name\"])\\n',\n",
       "  '        if not (self.plain_output):\\n',\n",
       "  '            self.output_writer.write_definition(self.conf)\\n',\n",
       "  '            self.output_writer.write_meta_line(\\n',\n",
       "  '                \"no_aggregate\", \",\".join(skip_aggregation)\\n',\n",
       "  '            )\\n',\n",
       "  '\\n',\n",
       "  '    def connect_db(self):\\n',\n",
       "  '        from pathlib import Path\\n',\n",
       "  '        import sqlite3\\n',\n",
       "  '\\n',\n",
       "  '        db_path = Path(self.data_dir) / (self.module_name + \".sqlite\")\\n',\n",
       "  '        if db_path.exists():\\n',\n",
       "  '            self.dbconn = sqlite3.connect(str(db_path))\\n',\n",
       "  '            self.cursor = self.dbconn.cursor()\\n',\n",
       "  '\\n',\n",
       "  '    def close_db_connection(self):\\n',\n",
       "  '        if self.cursor is not None:\\n',\n",
       "  '            self.cursor.close()\\n',\n",
       "  '        if self.dbconn is not None:\\n',\n",
       "  '            self.dbconn.close()\\n',\n",
       "  '\\n',\n",
       "  '    # Placeholder, intended to be overridded in derived class\\n',\n",
       "  '    def setup(self):\\n',\n",
       "  '        pass\\n',\n",
       "  '\\n',\n",
       "  '    def base_cleanup(self):\\n',\n",
       "  '        if self.output_writer:\\n',\n",
       "  '            self.output_writer.close()\\n',\n",
       "  '        # self.invalid_file.close()\\n',\n",
       "  '        if self.dbconn != None:\\n',\n",
       "  '            self.close_db_connection()\\n',\n",
       "  '        self.cleanup()\\n',\n",
       "  '\\n',\n",
       "  '    # Placeholder, intended to be overridden in derived class\\n',\n",
       "  '    def cleanup(self):\\n',\n",
       "  '        pass\\n',\n",
       "  '\\n',\n",
       "  '    def _setup_logger(self):\\n',\n",
       "  '        from logging import getLogger\\n',\n",
       "  '\\n',\n",
       "  '        self.logger = getLogger(\"oakvar.\" + self.module_name)\\n',\n",
       "  '        self.error_logger = getLogger(\"err.\" + self.module_name)\\n',\n",
       "  '\\n',\n",
       "  '    def _get_input(self):\\n',\n",
       "  '        from ..util.inout import AllMappingsParser\\n',\n",
       "  '        from ..consts import all_mappings_col_name\\n',\n",
       "  '        from ..consts import mapping_parser_name\\n',\n",
       "  '        from ..exceptions import SetupError\\n',\n",
       "  '\\n',\n",
       "  '        if self.conf is None or self.primary_input_reader is None:\\n',\n",
       "  '            raise SetupError(self.module_name)\\n',\n",
       "  '        for lnum, line, reader_data in self.primary_input_reader.loop_data():\\n',\n",
       "  '            try:\\n',\n",
       "  '                input_data = {}\\n',\n",
       "  '                for col_name in self.conf[\"input_columns\"]:\\n',\n",
       "  '                    input_data[col_name] = reader_data[col_name]\\n',\n",
       "  '                if all_mappings_col_name in input_data:\\n',\n",
       "  '                    input_data[mapping_parser_name] = AllMappingsParser(\\n',\n",
       "  '                        input_data[all_mappings_col_name]\\n',\n",
       "  '                    )\\n',\n",
       "  '                secondary_data = {}\\n',\n",
       "  '                for module_name, fetcher in self.secondary_readers.items():\\n',\n",
       "  '                    input_key_col = (\\n',\n",
       "  '                        self.conf[\"secondary_inputs\"][module_name]\\n',\n",
       "  '                        .get(\"match_columns\", {})\\n',\n",
       "  '                        .get(\"primary\", \"uid\")\\n',\n",
       "  '                    )\\n',\n",
       "  '                    input_key_data = input_data[input_key_col]\\n',\n",
       "  '                    secondary_data[module_name] = fetcher.get(input_key_data)\\n',\n",
       "  '                yield lnum, line, input_data, secondary_data\\n',\n",
       "  '            except Exception as e:\\n',\n",
       "  '                self._log_runtime_exception(\\n',\n",
       "  '                    lnum, line, reader_data, e, fn=self.primary_input_reader.path\\n',\n",
       "  '                )\\n',\n",
       "  '                continue\\n',\n",
       "  '\\n',\n",
       "  '    def annotate(self, input_data, secondary_data=None):\\n',\n",
       "  '        return {\\n',\n",
       "  '            \"default__annotation\": \"no annotate has been defined\",\\n',\n",
       "  '            \"input_data\": input_data,\\n',\n",
       "  '            \"secondary_data\": secondary_data,\\n',\n",
       "  '        }\\n',\n",
       "  '\\n',\n",
       "  '    def live_report_substitute(self, d):\\n',\n",
       "  '        import re\\n',\n",
       "  '        from ..exceptions import SetupError\\n',\n",
       "  '\\n',\n",
       "  '        if self.conf is None:\\n',\n",
       "  '            raise SetupError(self.module_name)\\n',\n",
       "  '        if \"report_substitution\" not in self.conf:\\n',\n",
       "  '            return d\\n',\n",
       "  '        rs_dic = self.conf[\"report_substitution\"]\\n',\n",
       "  '        rs_dic_keys = list(rs_dic.keys())\\n',\n",
       "  '        for colname in d.keys():\\n',\n",
       "  '            if colname in rs_dic_keys:\\n',\n",
       "  '                value = d[colname]\\n',\n",
       "  '                if colname in [\"all_mappings\", \"all_so\"]:\\n',\n",
       "  '                    for target in list(rs_dic[colname].keys()):\\n',\n",
       "  '                        value = re.sub(\\n',\n",
       "  '                            \"\\\\\\\\b\" + target + \"\\\\\\\\b\", rs_dic[colname][target], value\\n',\n",
       "  '                        )\\n',\n",
       "  '                else:\\n',\n",
       "  '                    if value in rs_dic[colname]:\\n',\n",
       "  '                        value = rs_dic[colname][value]\\n',\n",
       "  '                d[colname] = value\\n',\n",
       "  '        return d\\n',\n",
       "  '\\n',\n",
       "  '    def create_module_files(self):\\n',\n",
       "  '        pass\\n'],\n",
       " 6)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inspect.getsourcelines(ov.BaseAnnotator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "29a35132-a5b6-4bae-ba32-13cc04aecc31",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '__main__' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [38]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43m__main__\u001b[49m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__file__\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtest.py\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mNameError\u001b[0m: name '__main__' is not defined"
     ]
    }
   ],
   "source": [
    "__main__.__file__ = \"test.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7f6df2c3-dd18-43d2-8490-4a3be48c0ab5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'type'>\n"
     ]
    }
   ],
   "source": [
    "class A:\n",
    "    pass\n",
    "print(type(A))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6fc380e2-6173-41c0-86a9-365cdcb68a29",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "module, class, method, function, traceback, frame, or code object was expected, got Annotator",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [14]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01minspect\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43minspect\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetsource\u001b[49m\u001b[43m(\u001b[49m\u001b[43mm\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:997\u001b[0m, in \u001b[0;36mgetsource\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    991\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgetsource\u001b[39m(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    992\u001b[0m     \u001b[38;5;124;03m\"\"\"Return the text of the source code for an object.\u001b[39;00m\n\u001b[1;32m    993\u001b[0m \n\u001b[1;32m    994\u001b[0m \u001b[38;5;124;03m    The argument may be a module, class, method, function, traceback, frame,\u001b[39;00m\n\u001b[1;32m    995\u001b[0m \u001b[38;5;124;03m    or code object.  The source code is returned as a single string.  An\u001b[39;00m\n\u001b[1;32m    996\u001b[0m \u001b[38;5;124;03m    OSError is raised if the source code cannot be retrieved.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 997\u001b[0m     lines, lnum \u001b[38;5;241m=\u001b[39m \u001b[43mgetsourcelines\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mobject\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    998\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(lines)\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:979\u001b[0m, in \u001b[0;36mgetsourcelines\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    971\u001b[0m \u001b[38;5;124;03m\"\"\"Return a list of source lines and starting line number for an object.\u001b[39;00m\n\u001b[1;32m    972\u001b[0m \n\u001b[1;32m    973\u001b[0m \u001b[38;5;124;03mThe argument may be a module, class, method, function, traceback, frame,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    976\u001b[0m \u001b[38;5;124;03moriginal source file the first line of code was found.  An OSError is\u001b[39;00m\n\u001b[1;32m    977\u001b[0m \u001b[38;5;124;03mraised if the source code cannot be retrieved.\"\"\"\u001b[39;00m\n\u001b[1;32m    978\u001b[0m \u001b[38;5;28mobject\u001b[39m \u001b[38;5;241m=\u001b[39m unwrap(\u001b[38;5;28mobject\u001b[39m)\n\u001b[0;32m--> 979\u001b[0m lines, lnum \u001b[38;5;241m=\u001b[39m \u001b[43mfindsource\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mobject\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    981\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m istraceback(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    982\u001b[0m     \u001b[38;5;28mobject\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39mtb_frame\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:780\u001b[0m, in \u001b[0;36mfindsource\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    772\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfindsource\u001b[39m(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    773\u001b[0m     \u001b[38;5;124;03m\"\"\"Return the entire source file and starting line number for an object.\u001b[39;00m\n\u001b[1;32m    774\u001b[0m \n\u001b[1;32m    775\u001b[0m \u001b[38;5;124;03m    The argument may be a module, class, method, function, traceback, frame,\u001b[39;00m\n\u001b[1;32m    776\u001b[0m \u001b[38;5;124;03m    or code object.  The source code is returned as a list of all the lines\u001b[39;00m\n\u001b[1;32m    777\u001b[0m \u001b[38;5;124;03m    in the file and the line number indexes a line in that list.  An OSError\u001b[39;00m\n\u001b[1;32m    778\u001b[0m \u001b[38;5;124;03m    is raised if the source code cannot be retrieved.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 780\u001b[0m     file \u001b[38;5;241m=\u001b[39m \u001b[43mgetsourcefile\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mobject\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    781\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m file:\n\u001b[1;32m    782\u001b[0m         \u001b[38;5;66;03m# Invalidate cache if needed.\u001b[39;00m\n\u001b[1;32m    783\u001b[0m         linecache\u001b[38;5;241m.\u001b[39mcheckcache(file)\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:696\u001b[0m, in \u001b[0;36mgetsourcefile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    692\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgetsourcefile\u001b[39m(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    693\u001b[0m     \u001b[38;5;124;03m\"\"\"Return the filename that can be used to locate an object's source.\u001b[39;00m\n\u001b[1;32m    694\u001b[0m \u001b[38;5;124;03m    Return None if no way can be identified to get the source.\u001b[39;00m\n\u001b[1;32m    695\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 696\u001b[0m     filename \u001b[38;5;241m=\u001b[39m \u001b[43mgetfile\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mobject\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    697\u001b[0m     all_bytecode_suffixes \u001b[38;5;241m=\u001b[39m importlib\u001b[38;5;241m.\u001b[39mmachinery\u001b[38;5;241m.\u001b[39mDEBUG_BYTECODE_SUFFIXES[:]\n\u001b[1;32m    698\u001b[0m     all_bytecode_suffixes \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m importlib\u001b[38;5;241m.\u001b[39mmachinery\u001b[38;5;241m.\u001b[39mOPTIMIZED_BYTECODE_SUFFIXES[:]\n",
      "File \u001b[0;32m~/mambaforge/envs/ov/lib/python3.8/inspect.py:676\u001b[0m, in \u001b[0;36mgetfile\u001b[0;34m(object)\u001b[0m\n\u001b[1;32m    674\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m iscode(\u001b[38;5;28mobject\u001b[39m):\n\u001b[1;32m    675\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mobject\u001b[39m\u001b[38;5;241m.\u001b[39mco_filename\n\u001b[0;32m--> 676\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmodule, class, method, function, traceback, frame, or \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    677\u001b[0m                 \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcode object was expected, got \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(\n\u001b[1;32m    678\u001b[0m                 \u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mobject\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m))\n",
      "\u001b[0;31mTypeError\u001b[0m: module, class, method, function, traceback, frame, or code object was expected, got Annotator"
     ]
    }
   ],
   "source": [
    "import inspect\n",
    "inspect.getsource(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1361e85d-4346-4015-99fa-649e4b720268",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@ conf={'datasource': '2022.06.14', 'description': 'ClinVar is an archive of reports of the relationships among human variations and phenotypes, with supporting evidence.', 'developer': {'citation': 'Landrum, M.J., Lee, J.M., Benson, M., et al. (2016). ClinVar: public archive of interpretations of clinically relevant variants. Nucleic Acids Research, 44(D1):D862-8.', 'email': 'support@cravat.us', 'name': 'NCBI', 'organization': 'NCBI', 'website': 'https://www.ncbi.nlm.nih.gov/clinvar/'}, 'input_format': 'crx', 'level': 'variant', 'output_columns': [{'category': 'single', 'name': 'sig', 'title': 'Clinical Significance', 'type': 'string', 'width': 90, 'filterable': False, 'hidden': True}, {'hidden': True, 'name': 'disease_refs', 'title': 'Disease Ref Nums', 'type': 'string', 'width': 55, 'filterable': False, 'desc': 'Disease reference numbers'}, {'name': 'disease_names', 'title': 'Disease Names', 'type': 'string', 'width': 55, 'hidden': False}, {'desc': 'The level of review supporting the assertion of clinical significance', 'name': 'rev_stat', 'title': 'Review Status', 'type': 'string', 'width': 55, 'filterable': False, 'hidden': True}, {'filterable': False, 'hidden': True, 'name': 'id', 'title': 'ClinVar ID', 'type': 'string', 'width': 55, 'link_format': 'https://www.ncbi.nlm.nih.gov/clinvar/variation/${.*}/'}, {'desc': 'Addional detail on clinical significance used when it is conflicting', 'name': 'sig_conf', 'title': 'Significance Detail', 'type': 'string', 'width': 55, 'filterable': False, 'hidden': True}], 'release_note': {'2022.06.14': 'Added significance detail for conflicting variants.  Data updated.', '2021.10.01': 'data updated', '2021.08.04': 'data updated', '2021.05.01': 'data updated', '2021.04.01': 'data updated', '2021.03.03': 'data updated', '2021.02.04': 'data updated', '2021.01.12': 'data updated', '2020.11.02': 'data updated', '2020.02.11.1': 'strict ref/alt indels, not fuzzy', '2020.02.11': 'data update', '2019.03.261': 'added requires.', '2019.01.021': 'added phenotype tag.', '2019.01.022': 'updated data set, widget overhaul', '2019.01.023': 'changed id column format for web linking'}, 'tags': ['variants', 'clinical relevance', 'literature'], 'title': 'ClinVar', 'type': 'annotator', 'version': '2022.06.14.1', 'data_version': '2022.06.14', 'requires_oakvar': '>=2.5.0', 'requires': ['wgclinvar'], 'smartfilters': [{'name': 'hasdata', 'title': 'ClinVar', 'description': 'Only include variants with data from ClinVar', 'selector': {'type': 'empty'}, 'filter': {'column': 'clinvar__id', 'test': 'hasData'}}]}\n"
     ]
    }
   ],
   "source": [
    "c=ov.get_live_annotator(module_name=\"clinvar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9e38e5fc-85ea-4515-bd8e-4b5cd840026c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sig': 'Pathogenic',\n",
       " 'disease_refs': 'MedGen:CN517202',\n",
       " 'disease_names': 'not provided',\n",
       " 'rev_stat': 'criteria provided, single submitter',\n",
       " 'id': 200993,\n",
       " 'sig_conf': None}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c.annotate({\"chrom\": \"chr11\", \"pos\": 64810053, \"ref_base\": \"-\", \"alt_base\": \"A\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05c0f32e-3f97-48c7-974e-582058a842a6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
